{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c1548fbc",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import allel\n",
    "import re\n",
    "\n",
    "def load_vcf(vcf_path, metadata):\n",
    "    \"\"\"\n",
    "    Load VCF and filter poor-quality samples\n",
    "    \"\"\"\n",
    "    \n",
    "    sampleIDs = metadata.sampleID.to_list()\n",
    "    \n",
    "    # load vcf and get genotypes and positions\n",
    "    vcf = allel.read_vcf(vcf_path, fields='*')\n",
    "    samples = vcf['samples']\n",
    "    # keep only samples in qcpass metadata \n",
    "    sample_mask = np.isin(vcf['samples'], metadata.sampleID)\n",
    "    \n",
    "    # remove low quality samples \n",
    "    geno = allel.GenotypeArray(vcf['calldata/GT'])\n",
    "    geno = geno.compress(sample_mask, axis=1)\n",
    "    pos = vcf['variants/POS']\n",
    "    contig = vcf['variants/CHROM']\n",
    "    indel = vcf['variants/INDEL']\n",
    "    \n",
    "    # remove indels \n",
    "    geno = geno.compress(~indel, axis=0)\n",
    "    pos = pos[~indel]\n",
    "    contig = contig[~indel]\n",
    "    \n",
    "    return geno, pos, contig, samples[sample_mask], vcf['variants/REF'][~indel], vcf['variants/ALT'][~indel]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "907d8f16",
   "metadata": {
    "tags": [
     "remove-input",
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "metadata_path = '../../../results/config/metadata.qcpass.tsv'\n",
    "kdr_marker_snps_path = '../../../resources/ag-vampir/Kdr_marker_SNPs.csv'\n",
    "vcf_path = \"../../../results/vcfs/targets/ampseq-vigg-01.annot.vcf\"\n",
    "cohort_cols = 'location,taxon'\n",
    "wkdir = \"../../..\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d83db822",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Load in the table of SNPs that we use to define haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "257966e1",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP name</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele1 interpretation</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Allele2 interpretation</th>\n",
       "      <th>variant_pos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP Position</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2L:2422651</th>\n",
       "      <td>kdr-995S</td>\n",
       "      <td>T</td>\n",
       "      <td>Not KdrS</td>\n",
       "      <td>C</td>\n",
       "      <td>KdrS</td>\n",
       "      <td>2422651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2422652</th>\n",
       "      <td>kdr-995F</td>\n",
       "      <td>A</td>\n",
       "      <td>Not KdrF</td>\n",
       "      <td>T</td>\n",
       "      <td>KdrF</td>\n",
       "      <td>2422652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2391228</th>\n",
       "      <td>kdr-402L</td>\n",
       "      <td>G</td>\n",
       "      <td>Not 402L</td>\n",
       "      <td>C or T</td>\n",
       "      <td>V402L</td>\n",
       "      <td>2391228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2403941</th>\n",
       "      <td>Def-S4S5</td>\n",
       "      <td>T</td>\n",
       "      <td>Not S4S5</td>\n",
       "      <td>C</td>\n",
       "      <td>kdr haplotype cluster S4 or S5</td>\n",
       "      <td>2403941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2414062</th>\n",
       "      <td>Def-F3F4-2</td>\n",
       "      <td>A</td>\n",
       "      <td>Not F3F4</td>\n",
       "      <td>T</td>\n",
       "      <td>kdr haplotype cluster F3 or F4</td>\n",
       "      <td>2414062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2399997</th>\n",
       "      <td>Def-F3</td>\n",
       "      <td>G</td>\n",
       "      <td>Not F3</td>\n",
       "      <td>C</td>\n",
       "      <td>kdr haplotype cluster F3</td>\n",
       "      <td>2399997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2425052</th>\n",
       "      <td>Def-F2</td>\n",
       "      <td>G</td>\n",
       "      <td>Not F2</td>\n",
       "      <td>A</td>\n",
       "      <td>kdr haplotype cluster F2</td>\n",
       "      <td>2425052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2380982</th>\n",
       "      <td>Def-S1-3</td>\n",
       "      <td>T</td>\n",
       "      <td>Not S1</td>\n",
       "      <td>C</td>\n",
       "      <td>kdr haplotype cluster S1</td>\n",
       "      <td>2380982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2408677</th>\n",
       "      <td>Def-S5</td>\n",
       "      <td>A</td>\n",
       "      <td>Not S5</td>\n",
       "      <td>C</td>\n",
       "      <td>kdr haplotype cluster S5</td>\n",
       "      <td>2408677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2381706</th>\n",
       "      <td>Def-F5-2</td>\n",
       "      <td>A</td>\n",
       "      <td>Not F5</td>\n",
       "      <td>G</td>\n",
       "      <td>kdr haplotype cluster F5</td>\n",
       "      <td>2381706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2388595</th>\n",
       "      <td>Def-S4S5-2</td>\n",
       "      <td>G</td>\n",
       "      <td>Not S4S5</td>\n",
       "      <td>T</td>\n",
       "      <td>kdr haplotype cluster S4 or S5</td>\n",
       "      <td>2388595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2425766</th>\n",
       "      <td>Def-S2-4</td>\n",
       "      <td>T</td>\n",
       "      <td>Not S2</td>\n",
       "      <td>A</td>\n",
       "      <td>kdr haplotype cluster S2</td>\n",
       "      <td>2425766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2432419</th>\n",
       "      <td>Def-S3</td>\n",
       "      <td>T</td>\n",
       "      <td>Not S3</td>\n",
       "      <td>G</td>\n",
       "      <td>kdr haplotype cluster S3</td>\n",
       "      <td>2432419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2435581</th>\n",
       "      <td>Def-F1</td>\n",
       "      <td>G</td>\n",
       "      <td>Not F1</td>\n",
       "      <td>A</td>\n",
       "      <td>kdr haplotype cluster F1</td>\n",
       "      <td>2435581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2408423</th>\n",
       "      <td>Def-S2S4</td>\n",
       "      <td>C</td>\n",
       "      <td>Not S2S4</td>\n",
       "      <td>T</td>\n",
       "      <td>kdr haplotype cluster S2 or S4</td>\n",
       "      <td>2408423</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                SNP name Allele1 Allele1 interpretation Allele2  \\\n",
       "SNP Position                                                      \n",
       "2L:2422651      kdr-995S       T               Not KdrS       C   \n",
       "2L:2422652      kdr-995F       A               Not KdrF       T   \n",
       "2L:2391228      kdr-402L       G               Not 402L  C or T   \n",
       "2L:2403941      Def-S4S5       T               Not S4S5       C   \n",
       "2L:2414062    Def-F3F4-2       A               Not F3F4       T   \n",
       "2L:2399997        Def-F3       G                 Not F3       C   \n",
       "2L:2425052        Def-F2       G                 Not F2       A   \n",
       "2L:2380982      Def-S1-3       T                 Not S1       C   \n",
       "2L:2408677        Def-S5       A                 Not S5       C   \n",
       "2L:2381706      Def-F5-2       A                 Not F5       G   \n",
       "2L:2388595    Def-S4S5-2       G               Not S4S5       T   \n",
       "2L:2425766      Def-S2-4       T                 Not S2       A   \n",
       "2L:2432419        Def-S3       T                 Not S3       G   \n",
       "2L:2435581        Def-F1       G                 Not F1       A   \n",
       "2L:2408423      Def-S2S4       C               Not S2S4       T   \n",
       "\n",
       "                      Allele2 interpretation  variant_pos  \n",
       "SNP Position                                               \n",
       "2L:2422651                              KdrS      2422651  \n",
       "2L:2422652                              KdrF      2422652  \n",
       "2L:2391228                             V402L      2391228  \n",
       "2L:2403941    kdr haplotype cluster S4 or S5      2403941  \n",
       "2L:2414062    kdr haplotype cluster F3 or F4      2414062  \n",
       "2L:2399997          kdr haplotype cluster F3      2399997  \n",
       "2L:2425052          kdr haplotype cluster F2      2425052  \n",
       "2L:2380982          kdr haplotype cluster S1      2380982  \n",
       "2L:2408677          kdr haplotype cluster S5      2408677  \n",
       "2L:2381706          kdr haplotype cluster F5      2381706  \n",
       "2L:2388595    kdr haplotype cluster S4 or S5      2388595  \n",
       "2L:2425766          kdr haplotype cluster S2      2425766  \n",
       "2L:2432419          kdr haplotype cluster S3      2432419  \n",
       "2L:2435581          kdr haplotype cluster F1      2435581  \n",
       "2L:2408423    kdr haplotype cluster S2 or S4      2408423  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort_cols = cohort_cols.split(\",\")\n",
    "\n",
    "hap_def = pd.read_csv(kdr_marker_snps_path, sep = '\\t', index_col = 1)\n",
    "hap_def['variant_pos'] = hap_def.index.str.replace('.*:', '', regex = True).astype('int')\n",
    "hap_def"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88ea5248",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### The functions for making the kdr background calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8beaef71",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "# Determine kdr F origin for a genotype\n",
    "def _F_kdr_origin_gen(genotypes, clean = True):\n",
    "    if 'sample_name' in genotypes.index:\n",
    "        sample_name = genotypes['sample_name']\n",
    "    else:\n",
    "        sample_name = genotypes.name\n",
    "    # Check for the 995F mutations\n",
    "    if pd.isnull(genotypes['kdr-995F']):\n",
    "        kdr_F_origins = 'F:unknown'\n",
    "    elif genotypes['kdr-995F'] == 'AA':\n",
    "        kdr_F_origins = 'F:wt_hom'\n",
    "    elif genotypes['kdr-995F'] == 'AT':\n",
    "        kdr_F_origins = 'F:het'\n",
    "    elif genotypes['kdr-995F'] == 'TT':\n",
    "        kdr_F_origins = 'F:hom'\n",
    "    else:\n",
    "        print(f'Unexpected kdr F genotype. {sample_name} {genotypes[\"kdr-995F\"]}')\n",
    "        kdr_F_origins = 'Fail. Unexpected kdr F genotype'\n",
    "    # If the individual has Fkdr, find out its origins\n",
    "    # For F homozygotes\n",
    "    if kdr_F_origins == 'F:hom':\n",
    "        if pd.isnull(genotypes['Def-F1']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1?'\n",
    "        elif genotypes['Def-F1'] == 'AA':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1_hom'\n",
    "        elif genotypes['Def-F1'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2?'\n",
    "        elif genotypes['Def-F2'] == 'AA':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2_hom'\n",
    "        elif genotypes['Def-F2'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F3F4-2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F3F4?'\n",
    "        elif genotypes['Def-F3F4-2'] == 'TT':\n",
    "            if pd.isnull(genotypes['Def-F3']):\n",
    "                kdr_F_origins = f'{kdr_F_origins},(F3F4)_hom'\n",
    "            elif genotypes['Def-F3'] == 'CC':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F3_hom'\n",
    "            elif genotypes['Def-F3'] == 'CG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F3_het,F4_het'\n",
    "            elif genotypes['Def-F3'] == 'GG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F4_hom'\n",
    "        elif genotypes['Def-F3F4-2'] == 'AT':\n",
    "            if pd.isnull(genotypes['Def-F3']):\n",
    "                kdr_F_origins = f'{kdr_F_origins},(F3F4)_het'\n",
    "            elif genotypes['Def-F3'] == 'CC':\n",
    "                kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F3F4, but homozygote for F3.'\n",
    "            elif genotypes['Def-F3'] == 'CG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F3_het'\n",
    "            elif genotypes['Def-F3'] == 'GG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F4_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F5-2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5?'\n",
    "        elif genotypes['Def-F5-2'] == 'GG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5_hom'\n",
    "        elif genotypes['Def-F5-2'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5_het'\n",
    "    # for F heterozygotes\n",
    "    elif kdr_F_origins == 'F:het':\n",
    "        if pd.isnull(genotypes['Def-F1']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1?'\n",
    "        elif genotypes['Def-F1'] == 'AA':\n",
    "            kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr, but homozygote for F1.'\n",
    "        elif genotypes['Def-F1'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2?'\n",
    "        elif genotypes['Def-F2'] == 'AA':\n",
    "            kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr, but homozygote for F2.'\n",
    "        elif genotypes['Def-F2'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F3F4-2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F3F4?'\n",
    "        elif genotypes['Def-F3F4-2'] == 'TT':\n",
    "            kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr, but homozygote for F3F4.'\n",
    "        elif genotypes['Def-F3F4-2'] == 'AT':\n",
    "            if pd.isnull(genotypes['Def-F3']):\n",
    "                kdr_F_origins = f'{kdr_F_origins},(F3F4)_het'\n",
    "            elif genotypes['Def-F3'] == 'CC':\n",
    "                kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr and F3F4, but homozygote for F3.'\n",
    "            elif genotypes['Def-F3'] == 'CG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F3_het'\n",
    "            elif genotypes['Def-F3'] == 'GG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F4_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F5-2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5?'\n",
    "        elif genotypes['Def-F5-2'] == 'GG':\n",
    "            kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr, but homozygote for F5.'\n",
    "        elif genotypes['Def-F5-2'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5_het'\n",
    "    if clean:\n",
    "        return(_kdr_gen_cleanup(kdr_F_origins))\n",
    "    else:\n",
    "        return(kdr_F_origins)\n",
    "\n",
    "\n",
    "# Determine kdr S origin for a genotype\n",
    "def _S_kdr_origin_gen(genotypes, clean = True, alternate_S4S5 = False):\n",
    "    if 'sample_name' in genotypes.index:\n",
    "        sample_name = genotypes['sample_name']\n",
    "    else:\n",
    "        sample_name = genotypes.name\n",
    "    # Check for the 995S mutations\n",
    "    if pd.isnull(genotypes['kdr-995S']):\n",
    "        kdr_S_origins = 'S:unknown'\n",
    "    elif genotypes['kdr-995S'] == 'TT':\n",
    "        kdr_S_origins = 'S:wt_hom'\n",
    "    elif genotypes['kdr-995S'] == 'CT':\n",
    "        kdr_S_origins = 'S:het'\n",
    "    elif genotypes['kdr-995S'] == 'CC':\n",
    "        kdr_S_origins = 'S:hom'\n",
    "    else:\n",
    "        print(f'Unexpected kdr S genotype. {sample_name} {genotypes[\"kdr-995S\"]}')\n",
    "        kdr_S_origins = 'Fail. Unexpected kdr S genotype'\n",
    "    # If the individual has Skdr, find out its origins\n",
    "    # For S homozygotes\n",
    "    if kdr_S_origins == 'S:hom':\n",
    "        if pd.isnull(genotypes['Def-S1-3']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1?'\n",
    "        elif genotypes['Def-S1-3'] == 'CC':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1_hom'\n",
    "        elif genotypes['Def-S1-3'] == 'CT':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-S2S4']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S2S4?'\n",
    "        elif genotypes['Def-S2S4'] == 'TT':\n",
    "            if pd.isnull(genotypes['Def-S2-4']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},(S2S4)_hom'\n",
    "            elif genotypes['Def-S2-4'] == 'AA':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S2_hom'\n",
    "            elif genotypes['Def-S2-4'] == 'AT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S2_het,S4_het'\n",
    "            elif genotypes['Def-S2-4'] == 'TT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4_hom'\n",
    "        elif genotypes['Def-S2S4'] == 'CT':\n",
    "            if pd.isnull(genotypes['Def-S2-4']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},(S2S4)_het'\n",
    "            elif genotypes['Def-S2-4'] == 'AA':\n",
    "                kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S2S4, but homozygote for S2.'\n",
    "            elif genotypes['Def-S2-4'] == 'AT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S2_het'\n",
    "            elif genotypes['Def-S2-4'] == 'TT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-S3']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3?'\n",
    "        elif genotypes['Def-S3'] == 'GG':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3_hom'\n",
    "        elif genotypes['Def-S3'] == 'GT':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3_het'\n",
    "        # \n",
    "        if alternate_S4S5:\n",
    "            if pd.isnull(genotypes['Def-S4S5-2']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4S5?'\n",
    "            elif genotypes['Def-S4S5-2'] == 'TT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_hom'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_hom'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het,S4_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_hom'\n",
    "            elif genotypes['Def-S4S5-2'] == 'GT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_het'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S4S5, but homozygote for S5.'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "        else :\n",
    "            if pd.isnull(genotypes['Def-S4S5']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4S5?'\n",
    "            elif genotypes['Def-S4S5'] == 'CC':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_hom'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_hom'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het,S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_hom'\n",
    "            elif genotypes['Def-S4S5'] == 'CT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_het'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S4S5, but homozygote for S5.'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "    # for S heterozygotes\n",
    "    elif kdr_S_origins == 'S:het':\n",
    "        if pd.isnull(genotypes['Def-S1-3']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1?'\n",
    "        elif genotypes['Def-S1-3'] == 'CC':\n",
    "            kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S1.'\n",
    "        elif genotypes['Def-S1-3'] == 'CT':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-S2S4']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S2S4?'\n",
    "        elif genotypes['Def-S2S4'] == 'TT':\n",
    "            kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S2S4.'\n",
    "        elif genotypes['Def-S2S4'] == 'CT':\n",
    "            if pd.isnull(genotypes['Def-S2-4']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},(S2S4)_het'\n",
    "            elif genotypes['Def-S2-4'] == 'AA':\n",
    "                ksr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr and S2S4, but homozygote for S2.'\n",
    "            elif genotypes['Def-S2-4'] == 'AT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S2_het'\n",
    "            elif genotypes['Def-S2-4'] == 'TT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-S3']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3?'\n",
    "        elif genotypes['Def-S3'] == 'GG':\n",
    "            kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S3.'\n",
    "        elif genotypes['Def-S3'] == 'GT':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3_het'\n",
    "        # \n",
    "        if alternate_S4S5:\n",
    "            if pd.isnull(genotypes['Def-S4S5_2']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4S5?'\n",
    "            elif genotypes['Def-S4S5-2'] == 'TT':\n",
    "                kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S4S5.'\n",
    "            elif genotypes['Def-S4S5-2'] == 'GT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_het'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr and S4S5, but homozygote for S5.'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "        else :\n",
    "            if pd.isnull(genotypes['Def-S4S5']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4S5?'\n",
    "            elif genotypes['Def-S4S5'] == 'CC':\n",
    "                kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S4S5.'\n",
    "            elif genotypes['Def-S4S5'] == 'CT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_het'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    return(f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr and S4S5, but homozygote for S5.')\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "    if clean:\n",
    "        return(_kdr_gen_cleanup(kdr_S_origins))\n",
    "    else:\n",
    "        return(kdr_S_origins)\n",
    "\n",
    "def _402_kdr_origin_gen(genotypes, clean = True):\n",
    "    if 'sample_name' in genotypes.index:\n",
    "        sample_name = genotypes['sample_name']\n",
    "    else:\n",
    "        sample_name = genotypes.name\n",
    "    # Check for the 995F mutations\n",
    "    if pd.isnull(genotypes['kdr-402L']):\n",
    "        kdr_402_origins = '402:unknown'\n",
    "    elif genotypes['kdr-402L'] == 'GG':\n",
    "        kdr_402_origins = '402:wt_hom'\n",
    "    elif genotypes['kdr-402L'] == 'CG':\n",
    "        kdr_402_origins = '402:het,402LC_het'\n",
    "    elif genotypes['kdr-402L'] == 'GT':\n",
    "        kdr_402_origins = '402:het,402LT_het'\n",
    "    elif genotypes['kdr-402L'] == 'CT':\n",
    "        kdr_402_origins = '402:hom,402LC_het,402LT_het'\n",
    "    elif genotypes['kdr-402L'] == 'CC':\n",
    "        kdr_402_origins = '402:hom,402LC_hom'\n",
    "    elif genotypes['kdr-402L'] == 'TT':\n",
    "        kdr_402_origins = '402:hom,402LT_hom'\n",
    "    else:\n",
    "        print(f'Unexpected kdr 402 genotype. {sample_name} {genotypes[\"kdr-402L\"]}')\n",
    "        kdr_402_origins = 'Fail. Unexpected kdr 402 genotype'\n",
    "        \n",
    "    if clean:\n",
    "        return(_kdr_gen_cleanup(kdr_402_origins))\n",
    "    else:\n",
    "        return(kdr_402_origins)\n",
    "\n",
    "# The initial output of the kdr_origin function can be a little messy, since it outputs all of the \n",
    "# information that it could or couldn't obtain. This function tidies it up a bit. \n",
    "def _kdr_gen_cleanup(kdr_origin_str):\n",
    "    if re.search('Fail', kdr_origin_str):\n",
    "        return('?,?')\n",
    "    if re.search('wt', kdr_origin_str):\n",
    "        return('wt,wt')\n",
    "    kdr_type = re.findall('.*(?=:)', kdr_origin_str)[0]\n",
    "    outcomes = kdr_origin_str.split(',')\n",
    "    origins = np.unique(outcomes[1:])\n",
    "    established_origins = [o for o in origins if not re.search('\\?', o)]\n",
    "    # Remove \"_het\" and \"_hom\" text\n",
    "    established_origins = [re.sub('_het', '', o) for o in established_origins]\n",
    "    # Now double up each _hom entry. this is cludgy, but coudn't find a more elegant way\n",
    "    for i in range(len(established_origins)):\n",
    "        o = established_origins[i]\n",
    "        if re.search('_hom', o):\n",
    "            o = re.sub('_hom', '', o)\n",
    "            established_origins[i] = o\n",
    "            established_origins.append(o)\n",
    "    if re.search('hom', outcomes[0]):\n",
    "        if len(established_origins) == 1:\n",
    "            return(f'{kdr_type},{established_origins[0]}')\n",
    "        elif len(established_origins) == 2:\n",
    "            return(','.join(established_origins))\n",
    "        else:\n",
    "            return(f'{kdr_type},{kdr_type}')\n",
    "    if re.search('het', outcomes[0]):\n",
    "        if len(established_origins) == 1:\n",
    "            return(f'wt,{established_origins[0]}')\n",
    "        else:\n",
    "            return(f'wt,{kdr_type}')\n",
    "    else:\n",
    "        return('?,?')\n",
    "\n",
    "\n",
    "# Single function to call both the F and S origins for a given haplotype. This function determines\n",
    "# from the look of the genotype table whether it represents genotypes or haplotypes, and calls \n",
    "# the appropriate function. \n",
    "def kdr_origin(genotypes, alternate_S4S5 = False, clean = True, include_402 = None):\n",
    "    if 'sample_name' in genotypes.index:\n",
    "        sample_name = genotypes['sample_name']\n",
    "    else:\n",
    "        sample_name = genotypes.name\n",
    "    if include_402 == None:\n",
    "        if 'kdr-402L' in genotypes.index:\n",
    "            include_402 = True\n",
    "        else:\n",
    "            include_402 = False\n",
    "    if include_402 == False:\n",
    "        kdr_origins = pd.DataFrame({'kdr_F_origin': [_F_kdr_origin_gen(genotypes, clean)], \n",
    "                                    'kdr_S_origin': [_S_kdr_origin_gen(genotypes, clean, alternate_S4S5)]\n",
    "                                    }, index = [sample_name]\n",
    "        \n",
    "        )\n",
    "    else:\n",
    "        kdr_origins = pd.DataFrame({'kdr_F_origin': [_F_kdr_origin_gen(genotypes, clean)], \n",
    "                                    'kdr_S_origin': [_S_kdr_origin_gen(genotypes, clean, alternate_S4S5)],\n",
    "                                    'kdr_402_origin': [_402_kdr_origin_gen(genotypes, clean)]\n",
    "                                    }, index = [sample_name]\n",
    "        )\n",
    "    return(kdr_origins)\n",
    "\n",
    "# From a pair of kdr origin calls (obtained by running the kdr_origin function  for each of \n",
    "# F and S, followed by kdr_hap_cleanup), output a single call combining the F and S calls. \n",
    "def get_single_gen_call(x):  \n",
    "    if 'kdr_402_origin' in x.index:\n",
    "        return(_get_single_gen_call_with_402(x))\n",
    "    else:\n",
    "        return(_get_single_gen_call_no_402(x))\n",
    "    \n",
    "def _get_single_gen_call_no_402(x): \n",
    "    if 'sample_name' in x.index:\n",
    "        sample_name = x['sample_name']\n",
    "    else:\n",
    "        sample_name = x.name\n",
    "    joined_calls = np.array(x['kdr_F_origin'].split(',') + x['kdr_S_origin'].split(','))\n",
    "    # There should be at least two 'wt' calls\n",
    "    if np.sum(joined_calls == 'wt') < 2:\n",
    "        print(f'Too many different mutant haplotype backgrounds in sample {sample_name}')\n",
    "        return('?,?')\n",
    "    # Otherwise, drop two wildtype calls\n",
    "    else:\n",
    "        which_drop = np.where(joined_calls == 'wt')[0][:2]\n",
    "        return(','.join(np.delete(joined_calls, which_drop)))\n",
    "\n",
    "def _get_single_gen_call_with_402(x): \n",
    "    if 'sample_name' in x.index:\n",
    "        sample_name = x['sample_name']\n",
    "    else:\n",
    "        sample_name = x.name\n",
    "    joined_calls = np.array(x['kdr_F_origin'].split(',') + \n",
    "                            x['kdr_S_origin'].split(',') +\n",
    "                            x['kdr_402_origin'].split(',')\n",
    "    )\n",
    "    # There should be at least four 'wt' calls\n",
    "    if np.sum(np.isin(joined_calls,  ['wt', '?'])) < 4:\n",
    "        print(f'Too many different mutant haplotype backgrounds in sample {sample_name}')\n",
    "        return('?,?')\n",
    "    # Otherwise, drop four wildtype calls\n",
    "    else:\n",
    "        which_drop = np.concatenate([\n",
    "            np.where(joined_calls == 'wt')[0],\n",
    "            np.where(joined_calls == '?')[0]\n",
    "        ])[:4]\n",
    "        return(','.join(np.delete(joined_calls, which_drop)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81488cf0",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Load metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fd1a611c",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>taxon</th>\n",
       "      <th>location</th>\n",
       "      <th>country</th>\n",
       "      <th>plate</th>\n",
       "      <th>well_letter</th>\n",
       "      <th>well_number</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GH_01</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GH_02</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GH_04</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GH_05</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GH_06</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>665</th>\n",
       "      <td>Siaya_Delta_Alive_128</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>Siaya</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666</th>\n",
       "      <td>Siaya_Delta_Alive_129</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>Siaya</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>667</th>\n",
       "      <td>Siaya_Delta_Alive_130</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>Siaya</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>668</th>\n",
       "      <td>Siaya_Delta_Alive_131</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>Siaya</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>669</th>\n",
       "      <td>ML_17</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>ml</td>\n",
       "      <td>ml</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>596 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sampleID    taxon location country  plate well_letter  \\\n",
       "0                    GH_01     UNKN    Ghana   Ghana      3           A   \n",
       "1                    GH_02     UNKN    Ghana   Ghana      3           A   \n",
       "3                    GH_04     UNKN    Ghana   Ghana      3           A   \n",
       "4                    GH_05     UNKN    Ghana   Ghana      3           A   \n",
       "5                    GH_06     UNKN    Ghana   Ghana      3           A   \n",
       "..                     ...      ...      ...     ...    ...         ...   \n",
       "665  Siaya_Delta_Alive_128  gambiae    Siaya   Kenya      7           H   \n",
       "666  Siaya_Delta_Alive_129  gambiae    Siaya   Kenya      7           H   \n",
       "667  Siaya_Delta_Alive_130  gambiae    Siaya   Kenya      7           H   \n",
       "668  Siaya_Delta_Alive_131  gambiae    Siaya   Kenya      7           H   \n",
       "669                  ML_17     UNKN       ml      ml      7           H   \n",
       "\n",
       "     well_number  latitude  longitude  \n",
       "0              1       NaN        NaN  \n",
       "1              2       NaN        NaN  \n",
       "3              4       NaN        NaN  \n",
       "4              5       NaN        NaN  \n",
       "5              6       NaN        NaN  \n",
       "..           ...       ...        ...  \n",
       "665            6       NaN        NaN  \n",
       "666            7       NaN        NaN  \n",
       "667            8       NaN        NaN  \n",
       "668            9       NaN        NaN  \n",
       "669           10       NaN        NaN  \n",
       "\n",
       "[596 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv(metadata_path, sep = '\\t', index_col = 0)\n",
    "metadata.taxon.fillna('UNKN', inplace = True)\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ad48170",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Get the genotype calls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "507e6744",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Filter the SNPs to just the ones useful for kdr origin analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a5de2ea6",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sanj/.local/lib/python3.10/site-packages/allel/io/vcf_read.py:1732: UserWarning: invalid INFO header: '##INFO=<ID=VDB,Number=1,Type=Float,Description=\"Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)\",Version=\"3\">\\n'\n",
      "  warnings.warn('invalid INFO header: %r' % header)\n"
     ]
    }
   ],
   "source": [
    "geno, pos, contig, samples, ref, alt = load_vcf(vcf_path, metadata=metadata)\n",
    "\n",
    "which_snps = (contig == '2L') & np.isin(pos, hap_def['variant_pos']) \n",
    "\n",
    "snp_calls = geno[which_snps, :, :]\n",
    "pos = pos[which_snps]\n",
    "alt = alt[which_snps, :]\n",
    "ref = ref[which_snps]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2710ba21",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Convert genotype calls to nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5f235e86",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>SNP name</th>\n",
       "      <th>Def-S1-3</th>\n",
       "      <th>Def-F5-2</th>\n",
       "      <th>Def-S4S5-2</th>\n",
       "      <th>Def-F3</th>\n",
       "      <th>Def-S4S5</th>\n",
       "      <th>Def-S2S4</th>\n",
       "      <th>Def-S5</th>\n",
       "      <th>Def-F3F4-2</th>\n",
       "      <th>kdr-995S</th>\n",
       "      <th>kdr-995F</th>\n",
       "      <th>Def-F2</th>\n",
       "      <th>Def-S2-4</th>\n",
       "      <th>Def-S3</th>\n",
       "      <th>Def-F1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GH_01</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_02</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_04</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_05</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_06</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_128</th>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>CT</td>\n",
       "      <td>AT</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>??</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_129</th>\n",
       "      <td>CT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>CT</td>\n",
       "      <td>AT</td>\n",
       "      <td>AG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_130</th>\n",
       "      <td>CT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>CT</td>\n",
       "      <td>AT</td>\n",
       "      <td>AG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_131</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ML_17</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>??</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>596 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "SNP name              Def-S1-3 Def-F5-2 Def-S4S5-2 Def-F3 Def-S4S5 Def-S2S4  \\\n",
       "GH_01                       TT       AA         GG     GG       TT       CC   \n",
       "GH_02                       TT       AA         GG     GG       TT       CC   \n",
       "GH_04                       TT       AA         GG     GG       TT       CC   \n",
       "GH_05                       TT       AA         GG     GG       TT       CC   \n",
       "GH_06                       TT       AA         GG     GG       TT       CC   \n",
       "...                        ...      ...        ...    ...      ...      ...   \n",
       "Siaya_Delta_Alive_128       CC       AA         GG     GG       TT       CC   \n",
       "Siaya_Delta_Alive_129       CT       AA         GG     GG       TT       CC   \n",
       "Siaya_Delta_Alive_130       CT       AA         GG     GG       TT       CC   \n",
       "Siaya_Delta_Alive_131       TT       AA         GG     GG       TT       CC   \n",
       "ML_17                       TT       AA         GG     ??       TT       CC   \n",
       "\n",
       "SNP name              Def-S5 Def-F3F4-2 kdr-995S kdr-995F Def-F2 Def-S2-4  \\\n",
       "GH_01                     AA         AA       TT       TT     GG       TT   \n",
       "GH_02                     AA         AA       TT       TT     GG       TT   \n",
       "GH_04                     AA         AA       TT       TT     GG       TT   \n",
       "GH_05                     AA         AA       TT       TT     GG       TT   \n",
       "GH_06                     AA         AA       TT       TT     GG       TT   \n",
       "...                      ...        ...      ...      ...    ...      ...   \n",
       "Siaya_Delta_Alive_128     AA         AA       CT       AT     AA       TT   \n",
       "Siaya_Delta_Alive_129     AA         AA       CT       AT     AG       TT   \n",
       "Siaya_Delta_Alive_130     AA         AA       CT       AT     AG       TT   \n",
       "Siaya_Delta_Alive_131     AA         AA       TT       TT     AA       TT   \n",
       "ML_17                     AA         AA       CC       AA     AA       TT   \n",
       "\n",
       "SNP name              Def-S3 Def-F1  \n",
       "GH_01                     TT     AA  \n",
       "GH_02                     TT     AA  \n",
       "GH_04                     TT     AA  \n",
       "GH_05                     TT     AA  \n",
       "GH_06                     TT     AA  \n",
       "...                      ...    ...  \n",
       "Siaya_Delta_Alive_128     ??     GG  \n",
       "Siaya_Delta_Alive_129     TT     GG  \n",
       "Siaya_Delta_Alive_130     TT     GG  \n",
       "Siaya_Delta_Alive_131     TT     GG  \n",
       "ML_17                     GG     GG  \n",
       "\n",
       "[596 rows x 14 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine ref and alt into a single matrix, and add a column of '?' at the end, so that \n",
    "# any genotype call of -1 (missing) draws the '?' character'\n",
    "snp_alleles = np.concatenate([np.reshape(ref, (len(ref), 1)), \n",
    "                              alt,\n",
    "                              np.full((len(ref), 1), '?')], \n",
    "                             axis = 1)\n",
    "\n",
    "# Convert numberic calls to nucleotides, and sort each pair of nucleotides alphabetically\n",
    "# (so, eg, the genotype 'TA' becomes 'AT')\n",
    "snp_genotypes_3d = snp_alleles[\n",
    "    np.array(np.arange(snp_alleles.shape[0])).reshape(snp_alleles.shape[0], 1, 1), \n",
    "    snp_calls\n",
    "]\n",
    "snp_genotypes = np.apply_along_axis(lambda x: ''.join(np.sort(x)), 2, snp_genotypes_3d)\n",
    "\n",
    "# Store results in data frame\n",
    "hap_def.index = hap_def['variant_pos']\n",
    "gen_df = pd.DataFrame(\n",
    "    np.transpose(snp_genotypes), \n",
    "    index = samples,\n",
    "    columns = hap_def.loc[pos, 'SNP name']\n",
    ")\n",
    "gen_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb8620cc",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Obtain kdr origin calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b8eb15c6",
   "metadata": {
    "scrolled": true,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unexpected kdr F genotype. GM_21 ??\n",
      "Unexpected kdr S genotype. GM_21 ??\n",
      "Unexpected kdr F genotype. GM_27 ??\n",
      "Unexpected kdr S genotype. GM_27 ??\n",
      "Unexpected kdr F genotype. GM_56 ??\n",
      "Unexpected kdr S genotype. GM_56 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Alive_15 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Alive_15 ??\n",
      "Too many different mutant haplotype backgrounds in sample GM_21\n",
      "Too many different mutant haplotype backgrounds in sample GM_27\n",
      "Too many different mutant haplotype backgrounds in sample GM_56\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_03\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_38\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_39\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_47\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_58\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_70\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_71\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_73\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_15\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_33\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_66\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_69\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_80\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_85\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_87\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_89\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_108\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_114\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_109\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_119\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_128\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kdr_F_origin</th>\n",
       "      <th>kdr_S_origin</th>\n",
       "      <th>kdr_origin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GH_01</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_02</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_04</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_05</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_06</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_128</th>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_129</th>\n",
       "      <td>wt,F2</td>\n",
       "      <td>wt,S1</td>\n",
       "      <td>F2,S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_130</th>\n",
       "      <td>wt,F2</td>\n",
       "      <td>wt,S1</td>\n",
       "      <td>F2,S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_131</th>\n",
       "      <td>F2,F2</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F2,F2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ML_17</th>\n",
       "      <td>wt,wt</td>\n",
       "      <td>S3,S3</td>\n",
       "      <td>S3,S3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>596 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      kdr_F_origin kdr_S_origin kdr_origin\n",
       "GH_01                        F1,F1        wt,wt      F1,F1\n",
       "GH_02                        F1,F1        wt,wt      F1,F1\n",
       "GH_04                        F1,F1        wt,wt      F1,F1\n",
       "GH_05                        F1,F1        wt,wt      F1,F1\n",
       "GH_06                        F1,F1        wt,wt      F1,F1\n",
       "...                            ...          ...        ...\n",
       "Siaya_Delta_Alive_128          ?,?          ?,?        ?,?\n",
       "Siaya_Delta_Alive_129        wt,F2        wt,S1      F2,S1\n",
       "Siaya_Delta_Alive_130        wt,F2        wt,S1      F2,S1\n",
       "Siaya_Delta_Alive_131        F2,F2        wt,wt      F2,F2\n",
       "ML_17                        wt,wt        S3,S3      S3,S3\n",
       "\n",
       "[596 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "kdr_origins = pd.concat([kdr_origin(gen_df.iloc[i]) for i in range(gen_df.shape[0])])\n",
    "kdr_origins['kdr_origin'] = kdr_origins.apply(\n",
    "    get_single_gen_call, axis = 1\n",
    ")\n",
    "\n",
    "kdr_origins"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d31864b4",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Merge kdr origins with metadata and write to file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "22f47c9a",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kdr_F_origin</th>\n",
       "      <th>kdr_S_origin</th>\n",
       "      <th>kdr_origin</th>\n",
       "      <th>taxon</th>\n",
       "      <th>location</th>\n",
       "      <th>country</th>\n",
       "      <th>plate</th>\n",
       "      <th>well_letter</th>\n",
       "      <th>well_number</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GH_01</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_02</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_04</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_05</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_06</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_128</th>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>Siaya</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_129</th>\n",
       "      <td>wt,F2</td>\n",
       "      <td>wt,S1</td>\n",
       "      <td>F2,S1</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>Siaya</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_130</th>\n",
       "      <td>wt,F2</td>\n",
       "      <td>wt,S1</td>\n",
       "      <td>F2,S1</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>Siaya</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_131</th>\n",
       "      <td>F2,F2</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F2,F2</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>Siaya</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ML_17</th>\n",
       "      <td>wt,wt</td>\n",
       "      <td>S3,S3</td>\n",
       "      <td>S3,S3</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>ml</td>\n",
       "      <td>ml</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>596 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      kdr_F_origin kdr_S_origin kdr_origin    taxon location  \\\n",
       "GH_01                        F1,F1        wt,wt      F1,F1     UNKN    Ghana   \n",
       "GH_02                        F1,F1        wt,wt      F1,F1     UNKN    Ghana   \n",
       "GH_04                        F1,F1        wt,wt      F1,F1     UNKN    Ghana   \n",
       "GH_05                        F1,F1        wt,wt      F1,F1     UNKN    Ghana   \n",
       "GH_06                        F1,F1        wt,wt      F1,F1     UNKN    Ghana   \n",
       "...                            ...          ...        ...      ...      ...   \n",
       "Siaya_Delta_Alive_128          ?,?          ?,?        ?,?  gambiae    Siaya   \n",
       "Siaya_Delta_Alive_129        wt,F2        wt,S1      F2,S1  gambiae    Siaya   \n",
       "Siaya_Delta_Alive_130        wt,F2        wt,S1      F2,S1  gambiae    Siaya   \n",
       "Siaya_Delta_Alive_131        F2,F2        wt,wt      F2,F2  gambiae    Siaya   \n",
       "ML_17                        wt,wt        S3,S3      S3,S3     UNKN       ml   \n",
       "\n",
       "                      country  plate well_letter  well_number  latitude  \\\n",
       "GH_01                   Ghana      3           A            1       NaN   \n",
       "GH_02                   Ghana      3           A            2       NaN   \n",
       "GH_04                   Ghana      3           A            4       NaN   \n",
       "GH_05                   Ghana      3           A            5       NaN   \n",
       "GH_06                   Ghana      3           A            6       NaN   \n",
       "...                       ...    ...         ...          ...       ...   \n",
       "Siaya_Delta_Alive_128   Kenya      7           H            6       NaN   \n",
       "Siaya_Delta_Alive_129   Kenya      7           H            7       NaN   \n",
       "Siaya_Delta_Alive_130   Kenya      7           H            8       NaN   \n",
       "Siaya_Delta_Alive_131   Kenya      7           H            9       NaN   \n",
       "ML_17                      ml      7           H           10       NaN   \n",
       "\n",
       "                       longitude  \n",
       "GH_01                        NaN  \n",
       "GH_02                        NaN  \n",
       "GH_04                        NaN  \n",
       "GH_05                        NaN  \n",
       "GH_06                        NaN  \n",
       "...                          ...  \n",
       "Siaya_Delta_Alive_128        NaN  \n",
       "Siaya_Delta_Alive_129        NaN  \n",
       "Siaya_Delta_Alive_130        NaN  \n",
       "Siaya_Delta_Alive_131        NaN  \n",
       "ML_17                        NaN  \n",
       "\n",
       "[596 rows x 11 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kdr_origins_df = pd.merge(kdr_origins, metadata.set_index(\"sampleID\"), left_index = True, right_index = True)\n",
    "kdr_origins_df.to_csv(f'{wkdir}/results/kdr-origins/kdr_origins.csv', sep = '\\t')\n",
    "kdr_origins_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e0ce4db",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Create a table where each row is a haplotype instead of a genotype (although these are genotype-based calls, so order within each sample will be random, so they can be used for, say, mapping, but not haplotype clustering). Write this table to file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8ac2402e",
   "metadata": {
    "scrolled": true,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kdr_origin</th>\n",
       "      <th>taxon</th>\n",
       "      <th>location</th>\n",
       "      <th>country</th>\n",
       "      <th>plate</th>\n",
       "      <th>well_letter</th>\n",
       "      <th>well_number</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Calvin_01</th>\n",
       "      <td>wt</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Calvin_01</th>\n",
       "      <td>wt</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Calvin_02</th>\n",
       "      <td>wt</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Calvin_02</th>\n",
       "      <td>wt</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Calvin_03</th>\n",
       "      <td>wt</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VK7_dead_42</th>\n",
       "      <td>F1</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>VK7</td>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>10</td>\n",
       "      <td>H</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>random_arabiensis</th>\n",
       "      <td>wt</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>random</td>\n",
       "      <td>random</td>\n",
       "      <td>7</td>\n",
       "      <td>C</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>random_arabiensis</th>\n",
       "      <td>wt</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>random</td>\n",
       "      <td>random</td>\n",
       "      <td>7</td>\n",
       "      <td>C</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>random_coluzzii</th>\n",
       "      <td>F1</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>random</td>\n",
       "      <td>random</td>\n",
       "      <td>7</td>\n",
       "      <td>D</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>random_coluzzii</th>\n",
       "      <td>F1</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>random</td>\n",
       "      <td>random</td>\n",
       "      <td>7</td>\n",
       "      <td>D</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1192 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  kdr_origin       taxon location       country  plate  \\\n",
       "Calvin_01                 wt        UNKN     UNKN          UNKN     14   \n",
       "Calvin_01                 wt        UNKN     UNKN          UNKN     14   \n",
       "Calvin_02                 wt        UNKN     UNKN          UNKN     14   \n",
       "Calvin_02                 wt        UNKN     UNKN          UNKN     14   \n",
       "Calvin_03                 wt        UNKN     UNKN          UNKN     14   \n",
       "...                      ...         ...      ...           ...    ...   \n",
       "VK7_dead_42               F1    coluzzii      VK7  Burkina Faso     10   \n",
       "random_arabiensis         wt  arabiensis   random        random      7   \n",
       "random_arabiensis         wt  arabiensis   random        random      7   \n",
       "random_coluzzii           F1    coluzzii   random        random      7   \n",
       "random_coluzzii           F1    coluzzii   random        random      7   \n",
       "\n",
       "                  well_letter  well_number  latitude  longitude  \n",
       "Calvin_01                   A            1       NaN        NaN  \n",
       "Calvin_01                   A            1       NaN        NaN  \n",
       "Calvin_02                   A            2       NaN        NaN  \n",
       "Calvin_02                   A            2       NaN        NaN  \n",
       "Calvin_03                   A            3       NaN        NaN  \n",
       "...                       ...          ...       ...        ...  \n",
       "VK7_dead_42                 H            5       NaN        NaN  \n",
       "random_arabiensis           C           12       NaN        NaN  \n",
       "random_arabiensis           C           12       NaN        NaN  \n",
       "random_coluzzii             D           12       NaN        NaN  \n",
       "random_coluzzii             D           12       NaN        NaN  \n",
       "\n",
       "[1192 rows x 9 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "kdr_genhap_origins_df = pd.DataFrame({'kdr_origin': ','.join(list(kdr_origins['kdr_origin'])).split(',')},\n",
    "                                     index = np.repeat(kdr_origins.index, 2)\n",
    ")\n",
    "kdr_genhap_origins_df = pd.merge(kdr_genhap_origins_df, metadata.set_index(\"sampleID\"), left_index=True, right_index=True)\n",
    "kdr_genhap_origins_df.to_csv(f'{wkdir}/results/kdr-origins/kdr_genhap_origins.csv', sep = '\\t')\n",
    "kdr_genhap_origins_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dae4aae",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Summarise results by taxon and location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6d3b20e5",
   "metadata": {
    "scrolled": false,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counts of origins:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kdr_origin</th>\n",
       "      <th>F</th>\n",
       "      <th>F1</th>\n",
       "      <th>F2</th>\n",
       "      <th>S</th>\n",
       "      <th>S1</th>\n",
       "      <th>S3</th>\n",
       "      <th>wt</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>location</th>\n",
       "      <th>taxon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gambia</th>\n",
       "      <th>UNKN</th>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ghana</th>\n",
       "      <th>UNKN</th>\n",
       "      <td>0</td>\n",
       "      <td>154</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sante</th>\n",
       "      <th>UNKN</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya</th>\n",
       "      <th>gambiae</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>279</td>\n",
       "      <td>17</td>\n",
       "      <td>103</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNKN</th>\n",
       "      <th>UNKN</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VK7</th>\n",
       "      <th>coluzzii</th>\n",
       "      <td>1</td>\n",
       "      <td>171</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ml</th>\n",
       "      <th>UNKN</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">random</th>\n",
       "      <th>arabiensis</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coluzzii</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "kdr_origin           F   F1   F2   S   S1  S3   wt\n",
       "location taxon                                    \n",
       "Gambia   UNKN        5   11    0  11    0   0  129\n",
       "Ghana    UNKN        0  154    0   0    0   0    0\n",
       "Sante    UNKN        0    0    0   1    0   0   11\n",
       "Siaya    gambiae     3    0  279  17  103   2    6\n",
       "UNKN     UNKN        0    0    0   0    0   0  188\n",
       "VK7      coluzzii    1  171    0   0    0   0    0\n",
       "ml       UNKN        1    0    1   5    5  14    0\n",
       "random   arabiensis  0    0    0   0    0   0    2\n",
       "         coluzzii    0    2    0   0    0   0    0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Frequencies of known origins:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kdr_origin</th>\n",
       "      <th>F</th>\n",
       "      <th>F1</th>\n",
       "      <th>F2</th>\n",
       "      <th>S</th>\n",
       "      <th>S1</th>\n",
       "      <th>S3</th>\n",
       "      <th>wt</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>location</th>\n",
       "      <th>taxon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gambia</th>\n",
       "      <th>UNKN</th>\n",
       "      <td>0.0320</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ghana</th>\n",
       "      <th>UNKN</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sante</th>\n",
       "      <th>UNKN</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.083</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya</th>\n",
       "      <th>gambiae</th>\n",
       "      <td>0.0073</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.680</td>\n",
       "      <td>0.041</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNKN</th>\n",
       "      <th>UNKN</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VK7</th>\n",
       "      <th>coluzzii</th>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.990</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ml</th>\n",
       "      <th>UNKN</th>\n",
       "      <td>0.0380</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.190</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.5400</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">random</th>\n",
       "      <th>arabiensis</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coluzzii</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "kdr_origin                F     F1     F2      S    S1      S3     wt\n",
       "location taxon                                                       \n",
       "Gambia   UNKN        0.0320  0.071  0.000  0.071  0.00  0.0000  0.830\n",
       "Ghana    UNKN        0.0000  1.000  0.000  0.000  0.00  0.0000  0.000\n",
       "Sante    UNKN        0.0000  0.000  0.000  0.083  0.00  0.0000  0.920\n",
       "Siaya    gambiae     0.0073  0.000  0.680  0.041  0.25  0.0049  0.015\n",
       "UNKN     UNKN        0.0000  0.000  0.000  0.000  0.00  0.0000  1.000\n",
       "VK7      coluzzii    0.0058  0.990  0.000  0.000  0.00  0.0000  0.000\n",
       "ml       UNKN        0.0380  0.000  0.038  0.190  0.19  0.5400  0.000\n",
       "random   arabiensis  0.0000  0.000  0.000  0.000  0.00  0.0000  1.000\n",
       "         coluzzii    0.0000  1.000  0.000  0.000  0.00  0.0000  0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cols_keep = cohort_cols + ['kdr_origin']\n",
    "# Count the number of occurances of each haplotypes in each population\n",
    "# \"values\" could be any column that isn't specified elsewhere in the function. But it's \n",
    "# not allowed to be blank, so we had to pick one. \n",
    "pop_origin_counts = kdr_genhap_origins_df[cols_keep].pivot_table(columns='kdr_origin', \n",
    "                                                                    index=cohort_cols,\n",
    "                                                                    aggfunc=len\n",
    "                                                                   ).fillna(0).astype(int)\n",
    "\n",
    "# A function to round a number up to n_signif significant figures\n",
    "def signif(x, n_figs):\n",
    "    power = 10 ** np.floor(np.log10(np.abs(x).clip(1e-200)))\n",
    "    rounded = np.round(x / power, n_figs - 1) * power\n",
    "    return rounded\n",
    "\n",
    "# Calculate row totals of non-\"?\" columns\n",
    "if '?' in pop_origin_counts.columns:\n",
    "    pop_origin_counts = pop_origin_counts.drop('?', axis = 1)\n",
    "\n",
    "row_totals = pop_origin_counts.sum(axis = 1)\n",
    "# Calculate origin frequencies. We exclude the \"?\" calls for this\n",
    "pop_origin_freqs = pop_origin_counts.div(row_totals, axis = 0)\n",
    "# Round to 2 significant figures\n",
    "pop_origin_freqs = signif(pop_origin_freqs, 2)\n",
    "print('Counts of origins:')\n",
    "display(pop_origin_counts)\n",
    "print('\\n\\nFrequencies of known origins:')\n",
    "display(pop_origin_freqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e782b7c6",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### That's all folks"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "pythonGenomics",
   "language": "python",
   "name": "pythongenomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
