{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5197d6f3",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c1548fbc",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import allel\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "907d8f16",
   "metadata": {
    "tags": [
     "remove-input",
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "metadata_path = '../../../config/metadata.tsv'\n",
    "vcf_path = \"vcf\"\n",
    "cohort_column = 'location'\n",
    "wkdir = \"../..\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d83db822",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Load in the table of SNPs that we use to define haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "257966e1",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP name</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele1 interpretation</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Allele2 interpretation</th>\n",
       "      <th>variant_pos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP Position</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2L:2422651</th>\n",
       "      <td>kdr-995S</td>\n",
       "      <td>T</td>\n",
       "      <td>Not KdrS</td>\n",
       "      <td>C</td>\n",
       "      <td>KdrS</td>\n",
       "      <td>2422651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2422652</th>\n",
       "      <td>kdr-995F</td>\n",
       "      <td>A</td>\n",
       "      <td>Not KdrF</td>\n",
       "      <td>T</td>\n",
       "      <td>KdrF</td>\n",
       "      <td>2422652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2391228</th>\n",
       "      <td>kdr-402L</td>\n",
       "      <td>G</td>\n",
       "      <td>Not 402L</td>\n",
       "      <td>C or T</td>\n",
       "      <td>V402L</td>\n",
       "      <td>2391228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2403941</th>\n",
       "      <td>Def-S4S5</td>\n",
       "      <td>T</td>\n",
       "      <td>Not S4S5</td>\n",
       "      <td>C</td>\n",
       "      <td>kdr haplotype cluster S4 or S5</td>\n",
       "      <td>2403941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2414062</th>\n",
       "      <td>Def-F3F4-2</td>\n",
       "      <td>A</td>\n",
       "      <td>Not F3F4</td>\n",
       "      <td>T</td>\n",
       "      <td>kdr haplotype cluster F3 or F4</td>\n",
       "      <td>2414062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2399997</th>\n",
       "      <td>Def-F3</td>\n",
       "      <td>G</td>\n",
       "      <td>Not F3</td>\n",
       "      <td>C</td>\n",
       "      <td>kdr haplotype cluster F3</td>\n",
       "      <td>2399997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2425052</th>\n",
       "      <td>Def-F2</td>\n",
       "      <td>G</td>\n",
       "      <td>Not F2</td>\n",
       "      <td>A</td>\n",
       "      <td>kdr haplotype cluster F2</td>\n",
       "      <td>2425052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2380982</th>\n",
       "      <td>Def-S1-3</td>\n",
       "      <td>T</td>\n",
       "      <td>Not S1</td>\n",
       "      <td>C</td>\n",
       "      <td>kdr haplotype cluster S1</td>\n",
       "      <td>2380982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2408677</th>\n",
       "      <td>Def-S5</td>\n",
       "      <td>A</td>\n",
       "      <td>Not S5</td>\n",
       "      <td>C</td>\n",
       "      <td>kdr haplotype cluster S5</td>\n",
       "      <td>2408677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2381706</th>\n",
       "      <td>Def-F5-2</td>\n",
       "      <td>A</td>\n",
       "      <td>Not F5</td>\n",
       "      <td>G</td>\n",
       "      <td>kdr haplotype cluster F5</td>\n",
       "      <td>2381706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2388595</th>\n",
       "      <td>Def-S4S5-2</td>\n",
       "      <td>G</td>\n",
       "      <td>Not S4S5</td>\n",
       "      <td>T</td>\n",
       "      <td>kdr haplotype cluster S4 or S5</td>\n",
       "      <td>2388595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2425766</th>\n",
       "      <td>Def-S2-4</td>\n",
       "      <td>T</td>\n",
       "      <td>Not S2</td>\n",
       "      <td>A</td>\n",
       "      <td>kdr haplotype cluster S2</td>\n",
       "      <td>2425766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2432419</th>\n",
       "      <td>Def-S3</td>\n",
       "      <td>T</td>\n",
       "      <td>Not S3</td>\n",
       "      <td>G</td>\n",
       "      <td>kdr haplotype cluster S3</td>\n",
       "      <td>2432419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2435581</th>\n",
       "      <td>Def-F1</td>\n",
       "      <td>G</td>\n",
       "      <td>Not F1</td>\n",
       "      <td>A</td>\n",
       "      <td>kdr haplotype cluster F1</td>\n",
       "      <td>2435581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2L:2408423</th>\n",
       "      <td>Def-S2S4</td>\n",
       "      <td>C</td>\n",
       "      <td>Not S2S4</td>\n",
       "      <td>T</td>\n",
       "      <td>kdr haplotype cluster S2 or S4</td>\n",
       "      <td>2408423</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                SNP name Allele1 Allele1 interpretation Allele2  \\\n",
       "SNP Position                                                      \n",
       "2L:2422651      kdr-995S       T               Not KdrS       C   \n",
       "2L:2422652      kdr-995F       A               Not KdrF       T   \n",
       "2L:2391228      kdr-402L       G               Not 402L  C or T   \n",
       "2L:2403941      Def-S4S5       T               Not S4S5       C   \n",
       "2L:2414062    Def-F3F4-2       A               Not F3F4       T   \n",
       "2L:2399997        Def-F3       G                 Not F3       C   \n",
       "2L:2425052        Def-F2       G                 Not F2       A   \n",
       "2L:2380982      Def-S1-3       T                 Not S1       C   \n",
       "2L:2408677        Def-S5       A                 Not S5       C   \n",
       "2L:2381706      Def-F5-2       A                 Not F5       G   \n",
       "2L:2388595    Def-S4S5-2       G               Not S4S5       T   \n",
       "2L:2425766      Def-S2-4       T                 Not S2       A   \n",
       "2L:2432419        Def-S3       T                 Not S3       G   \n",
       "2L:2435581        Def-F1       G                 Not F1       A   \n",
       "2L:2408423      Def-S2S4       C               Not S2S4       T   \n",
       "\n",
       "                      Allele2 interpretation  variant_pos  \n",
       "SNP Position                                               \n",
       "2L:2422651                              KdrS      2422651  \n",
       "2L:2422652                              KdrF      2422652  \n",
       "2L:2391228                             V402L      2391228  \n",
       "2L:2403941    kdr haplotype cluster S4 or S5      2403941  \n",
       "2L:2414062    kdr haplotype cluster F3 or F4      2414062  \n",
       "2L:2399997          kdr haplotype cluster F3      2399997  \n",
       "2L:2425052          kdr haplotype cluster F2      2425052  \n",
       "2L:2380982          kdr haplotype cluster S1      2380982  \n",
       "2L:2408677          kdr haplotype cluster S5      2408677  \n",
       "2L:2381706          kdr haplotype cluster F5      2381706  \n",
       "2L:2388595    kdr haplotype cluster S4 or S5      2388595  \n",
       "2L:2425766          kdr haplotype cluster S2      2425766  \n",
       "2L:2432419          kdr haplotype cluster S3      2432419  \n",
       "2L:2435581          kdr haplotype cluster F1      2435581  \n",
       "2L:2408423    kdr haplotype cluster S2 or S4      2408423  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hap_def = pd.read_csv('Kdr_marker_SNPs.csv', sep = '\\t', index_col = 1)\n",
    "hap_def['variant_pos'] = hap_def.index.str.replace('.*:', '', regex = True).astype('int')\n",
    "hap_def"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88ea5248",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### The functions for making the kdr background calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8beaef71",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "# Determine kdr F origin for a genotype\n",
    "def _F_kdr_origin_gen(genotypes, clean = True):\n",
    "    if 'sample_name' in genotypes.index:\n",
    "        sample_name = genotypes['sample_name']\n",
    "    else:\n",
    "        sample_name = genotypes.name\n",
    "    # Check for the 995F mutations\n",
    "    if pd.isnull(genotypes['kdr-995F']):\n",
    "        kdr_F_origins = 'F:unknown'\n",
    "    elif genotypes['kdr-995F'] == 'AA':\n",
    "        kdr_F_origins = 'F:wt_hom'\n",
    "    elif genotypes['kdr-995F'] == 'AT':\n",
    "        kdr_F_origins = 'F:het'\n",
    "    elif genotypes['kdr-995F'] == 'TT':\n",
    "        kdr_F_origins = 'F:hom'\n",
    "    else:\n",
    "        print(f'Unexpected kdr F genotype. {sample_name} {genotypes[\"kdr-995F\"]}')\n",
    "        kdr_F_origins = 'Fail. Unexpected kdr F genotype'\n",
    "    # If the individual has Fkdr, find out its origins\n",
    "    # For F homozygotes\n",
    "    if kdr_F_origins == 'F:hom':\n",
    "        if pd.isnull(genotypes['Def-F1']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1?'\n",
    "        elif genotypes['Def-F1'] == 'AA':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1_hom'\n",
    "        elif genotypes['Def-F1'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2?'\n",
    "        elif genotypes['Def-F2'] == 'AA':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2_hom'\n",
    "        elif genotypes['Def-F2'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F3F4-2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F3F4?'\n",
    "        elif genotypes['Def-F3F4-2'] == 'TT':\n",
    "            if pd.isnull(genotypes['Def-F3']):\n",
    "                kdr_F_origins = f'{kdr_F_origins},(F3F4)_hom'\n",
    "            elif genotypes['Def-F3'] == 'CC':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F3_hom'\n",
    "            elif genotypes['Def-F3'] == 'CG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F3_het,F4_het'\n",
    "            elif genotypes['Def-F3'] == 'GG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F4_hom'\n",
    "        elif genotypes['Def-F3F4-2'] == 'AT':\n",
    "            if pd.isnull(genotypes['Def-F3']):\n",
    "                kdr_F_origins = f'{kdr_F_origins},(F3F4)_het'\n",
    "            elif genotypes['Def-F3'] == 'CC':\n",
    "                kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F3F4, but homozygote for F3.'\n",
    "            elif genotypes['Def-F3'] == 'CG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F3_het'\n",
    "            elif genotypes['Def-F3'] == 'GG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F4_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F5-2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5?'\n",
    "        elif genotypes['Def-F5-2'] == 'GG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5_hom'\n",
    "        elif genotypes['Def-F5-2'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5_het'\n",
    "    # for F heterozygotes\n",
    "    elif kdr_F_origins == 'F:het':\n",
    "        if pd.isnull(genotypes['Def-F1']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1?'\n",
    "        elif genotypes['Def-F1'] == 'AA':\n",
    "            kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr, but homozygote for F1.'\n",
    "        elif genotypes['Def-F1'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F1_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2?'\n",
    "        elif genotypes['Def-F2'] == 'AA':\n",
    "            kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr, but homozygote for F2.'\n",
    "        elif genotypes['Def-F2'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F2_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F3F4-2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F3F4?'\n",
    "        elif genotypes['Def-F3F4-2'] == 'TT':\n",
    "            kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr, but homozygote for F3F4.'\n",
    "        elif genotypes['Def-F3F4-2'] == 'AT':\n",
    "            if pd.isnull(genotypes['Def-F3']):\n",
    "                kdr_F_origins = f'{kdr_F_origins},(F3F4)_het'\n",
    "            elif genotypes['Def-F3'] == 'CC':\n",
    "                kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr and F3F4, but homozygote for F3.'\n",
    "            elif genotypes['Def-F3'] == 'CG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F3_het'\n",
    "            elif genotypes['Def-F3'] == 'GG':\n",
    "                kdr_F_origins = f'{kdr_F_origins},F4_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-F5-2']):\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5?'\n",
    "        elif genotypes['Def-F5-2'] == 'GG':\n",
    "            kdr_F_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for F kdr, but homozygote for F5.'\n",
    "        elif genotypes['Def-F5-2'] == 'AG':\n",
    "            kdr_F_origins = f'{kdr_F_origins},F5_het'\n",
    "    if clean:\n",
    "        return(_kdr_gen_cleanup(kdr_F_origins))\n",
    "    else:\n",
    "        return(kdr_F_origins)\n",
    "\n",
    "\n",
    "# Determine kdr S origin for a genotype\n",
    "def _S_kdr_origin_gen(genotypes, clean = True, alternate_S4S5 = False):\n",
    "    if 'sample_name' in genotypes.index:\n",
    "        sample_name = genotypes['sample_name']\n",
    "    else:\n",
    "        sample_name = genotypes.name\n",
    "    # Check for the 995S mutations\n",
    "    if pd.isnull(genotypes['kdr-995S']):\n",
    "        kdr_S_origins = 'S:unknown'\n",
    "    elif genotypes['kdr-995S'] == 'TT':\n",
    "        kdr_S_origins = 'S:wt_hom'\n",
    "    elif genotypes['kdr-995S'] == 'CT':\n",
    "        kdr_S_origins = 'S:het'\n",
    "    elif genotypes['kdr-995S'] == 'CC':\n",
    "        kdr_S_origins = 'S:hom'\n",
    "    else:\n",
    "        print(f'Unexpected kdr S genotype. {sample_name} {genotypes[\"kdr-995S\"]}')\n",
    "        kdr_S_origins = 'Fail. Unexpected kdr S genotype'\n",
    "    # If the individual has Skdr, find out its origins\n",
    "    # For S homozygotes\n",
    "    if kdr_S_origins == 'S:hom':\n",
    "        if pd.isnull(genotypes['Def-S1-3']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1?'\n",
    "        elif genotypes['Def-S1-3'] == 'CC':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1_hom'\n",
    "        elif genotypes['Def-S1-3'] == 'CT':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-S2S4']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S2S4?'\n",
    "        elif genotypes['Def-S2S4'] == 'TT':\n",
    "            if pd.isnull(genotypes['Def-S2-4']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},(S2S4)_hom'\n",
    "            elif genotypes['Def-S2-4'] == 'AA':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S2_hom'\n",
    "            elif genotypes['Def-S2-4'] == 'AT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S2_het,S4_het'\n",
    "            elif genotypes['Def-S2-4'] == 'TT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4_hom'\n",
    "        elif genotypes['Def-S2S4'] == 'CT':\n",
    "            if pd.isnull(genotypes['Def-S2-4']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},(S2S4)_het'\n",
    "            elif genotypes['Def-S2-4'] == 'AA':\n",
    "                kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S2S4, but homozygote for S2.'\n",
    "            elif genotypes['Def-S2-4'] == 'AT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S2_het'\n",
    "            elif genotypes['Def-S2-4'] == 'TT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-S3']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3?'\n",
    "        elif genotypes['Def-S3'] == 'GG':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3_hom'\n",
    "        elif genotypes['Def-S3'] == 'GT':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3_het'\n",
    "        # \n",
    "        if alternate_S4S5:\n",
    "            if pd.isnull(genotypes['Def-S4S5-2']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4S5?'\n",
    "            elif genotypes['Def-S4S5-2'] == 'TT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_hom'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_hom'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het,S4_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_hom'\n",
    "            elif genotypes['Def-S4S5-2'] == 'GT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_het'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S4S5, but homozygote for S5.'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "        else :\n",
    "            if pd.isnull(genotypes['Def-S4S5']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4S5?'\n",
    "            elif genotypes['Def-S4S5'] == 'CC':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_hom'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_hom'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het,S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_hom'\n",
    "            elif genotypes['Def-S4S5'] == 'CT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_het'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S4S5, but homozygote for S5.'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "    # for S heterozygotes\n",
    "    elif kdr_S_origins == 'S:het':\n",
    "        if pd.isnull(genotypes['Def-S1-3']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1?'\n",
    "        elif genotypes['Def-S1-3'] == 'CC':\n",
    "            kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S1.'\n",
    "        elif genotypes['Def-S1-3'] == 'CT':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S1_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-S2S4']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S2S4?'\n",
    "        elif genotypes['Def-S2S4'] == 'TT':\n",
    "            kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S2S4.'\n",
    "        elif genotypes['Def-S2S4'] == 'CT':\n",
    "            if pd.isnull(genotypes['Def-S2-4']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},(S2S4)_het'\n",
    "            elif genotypes['Def-S2-4'] == 'AA':\n",
    "                ksr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr and S2S4, but homozygote for S2.'\n",
    "            elif genotypes['Def-S2-4'] == 'AT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S2_het'\n",
    "            elif genotypes['Def-S2-4'] == 'TT':\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "        #\n",
    "        if pd.isnull(genotypes['Def-S3']):\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3?'\n",
    "        elif genotypes['Def-S3'] == 'GG':\n",
    "            kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S3.'\n",
    "        elif genotypes['Def-S3'] == 'GT':\n",
    "            kdr_S_origins = f'{kdr_S_origins},S3_het'\n",
    "        # \n",
    "        if alternate_S4S5:\n",
    "            if pd.isnull(genotypes['Def-S4S5_2']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4S5?'\n",
    "            elif genotypes['Def-S4S5-2'] == 'TT':\n",
    "                kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S4S5.'\n",
    "            elif genotypes['Def-S4S5-2'] == 'GT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_het'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr and S4S5, but homozygote for S5.'\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "        else :\n",
    "            if pd.isnull(genotypes['Def-S4S5']):\n",
    "                kdr_S_origins = f'{kdr_S_origins},S4S5?'\n",
    "            elif genotypes['Def-S4S5'] == 'CC':\n",
    "                kdr_S_origins = f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr, but homozygote for S4S5.'\n",
    "            elif genotypes['Def-S4S5'] == 'CT':\n",
    "                if pd.isnull(genotypes['Def-S5']):\n",
    "                    kdr_S_origins = f'{kdr_S_origins},(S4S5)_het'\n",
    "                elif genotypes['Def-S5'] == 'CC':\n",
    "                    return(f'Fail. Genotypes suggest that sample {sample_name} is heterozygote for S kdr and S4S5, but homozygote for S5.')\n",
    "                elif genotypes['Def-S5'] == 'AC':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S5_het'\n",
    "                elif genotypes['Def-S5'] == 'AA':\n",
    "                    kdr_S_origins = f'{kdr_S_origins},S4_het'\n",
    "    if clean:\n",
    "        return(_kdr_gen_cleanup(kdr_S_origins))\n",
    "    else:\n",
    "        return(kdr_S_origins)\n",
    "\n",
    "def _402_kdr_origin_gen(genotypes, clean = True):\n",
    "    if 'sample_name' in genotypes.index:\n",
    "        sample_name = genotypes['sample_name']\n",
    "    else:\n",
    "        sample_name = genotypes.name\n",
    "    # Check for the 995F mutations\n",
    "    if pd.isnull(genotypes['kdr-402L']):\n",
    "        kdr_402_origins = '402:unknown'\n",
    "    elif genotypes['kdr-402L'] == 'GG':\n",
    "        kdr_402_origins = '402:wt_hom'\n",
    "    elif genotypes['kdr-402L'] == 'CG':\n",
    "        kdr_402_origins = '402:het,402LC_het'\n",
    "    elif genotypes['kdr-402L'] == 'GT':\n",
    "        kdr_402_origins = '402:het,402LT_het'\n",
    "    elif genotypes['kdr-402L'] == 'CT':\n",
    "        kdr_402_origins = '402:hom,402LC_het,402LT_het'\n",
    "    elif genotypes['kdr-402L'] == 'CC':\n",
    "        kdr_402_origins = '402:hom,402LC_hom'\n",
    "    elif genotypes['kdr-402L'] == 'TT':\n",
    "        kdr_402_origins = '402:hom,402LT_hom'\n",
    "    else:\n",
    "        print(f'Unexpected kdr 402 genotype. {sample_name} {genotypes[\"kdr-402L\"]}')\n",
    "        kdr_402_origins = 'Fail. Unexpected kdr 402 genotype'\n",
    "        \n",
    "    if clean:\n",
    "        return(_kdr_gen_cleanup(kdr_402_origins))\n",
    "    else:\n",
    "        return(kdr_402_origins)\n",
    "\n",
    "# The initial output of the kdr_origin function can be a little messy, since it outputs all of the \n",
    "# information that it could or couldn't obtain. This function tidies it up a bit. \n",
    "def _kdr_gen_cleanup(kdr_origin_str):\n",
    "    if re.search('Fail', kdr_origin_str):\n",
    "        return('?,?')\n",
    "    if re.search('wt', kdr_origin_str):\n",
    "        return('wt,wt')\n",
    "    kdr_type = re.findall('.*(?=:)', kdr_origin_str)[0]\n",
    "    outcomes = kdr_origin_str.split(',')\n",
    "    origins = np.unique(outcomes[1:])\n",
    "    established_origins = [o for o in origins if not re.search('\\?', o)]\n",
    "    # Remove \"_het\" and \"_hom\" text\n",
    "    established_origins = [re.sub('_het', '', o) for o in established_origins]\n",
    "    # Now double up each _hom entry. this is cludgy, but coudn't find a more elegant way\n",
    "    for i in range(len(established_origins)):\n",
    "        o = established_origins[i]\n",
    "        if re.search('_hom', o):\n",
    "            o = re.sub('_hom', '', o)\n",
    "            established_origins[i] = o\n",
    "            established_origins.append(o)\n",
    "    if re.search('hom', outcomes[0]):\n",
    "        if len(established_origins) == 1:\n",
    "            return(f'{kdr_type},{established_origins[0]}')\n",
    "        elif len(established_origins) == 2:\n",
    "            return(','.join(established_origins))\n",
    "        else:\n",
    "            return(f'{kdr_type},{kdr_type}')\n",
    "    if re.search('het', outcomes[0]):\n",
    "        if len(established_origins) == 1:\n",
    "            return(f'wt,{established_origins[0]}')\n",
    "        else:\n",
    "            return(f'wt,{kdr_type}')\n",
    "    else:\n",
    "        return('?,?')\n",
    "\n",
    "\n",
    "# Single function to call both the F and S origins for a given haplotype. This function determines\n",
    "# from the look of the genotype table whether it represents genotypes or haplotypes, and calls \n",
    "# the appropriate function. \n",
    "def kdr_origin(genotypes, alternate_S4S5 = False, clean = True, include_402 = None):\n",
    "    if 'sample_name' in genotypes.index:\n",
    "        sample_name = genotypes['sample_name']\n",
    "    else:\n",
    "        sample_name = genotypes.name\n",
    "    if include_402 == None:\n",
    "        if 'kdr-402L' in genotypes.index:\n",
    "            include_402 = True\n",
    "        else:\n",
    "            include_402 = False\n",
    "    if include_402 == False:\n",
    "        kdr_origins = pd.DataFrame({'kdr_F_origin': [_F_kdr_origin_gen(genotypes, clean)], \n",
    "                                    'kdr_S_origin': [_S_kdr_origin_gen(genotypes, clean, alternate_S4S5)]\n",
    "                                    }, index = [sample_name]\n",
    "        \n",
    "        )\n",
    "    else:\n",
    "        kdr_origins = pd.DataFrame({'kdr_F_origin': [_F_kdr_origin_gen(genotypes, clean)], \n",
    "                                    'kdr_S_origin': [_S_kdr_origin_gen(genotypes, clean, alternate_S4S5)],\n",
    "                                    'kdr_402_origin': [_402_kdr_origin_gen(genotypes, clean)]\n",
    "                                    }, index = [sample_name]\n",
    "        )\n",
    "    return(kdr_origins)\n",
    "\n",
    "# From a pair of kdr origin calls (obtained by running the kdr_origin function  for each of \n",
    "# F and S, followed by kdr_hap_cleanup), output a single call combining the F and S calls. \n",
    "def get_single_gen_call(x):  \n",
    "    if 'kdr_402_origin' in x.index:\n",
    "        return(_get_single_gen_call_with_402(x))\n",
    "    else:\n",
    "        return(_get_single_gen_call_no_402(x))\n",
    "    \n",
    "def _get_single_gen_call_no_402(x): \n",
    "    if 'sample_name' in x.index:\n",
    "        sample_name = x['sample_name']\n",
    "    else:\n",
    "        sample_name = x.name\n",
    "    joined_calls = np.array(x['kdr_F_origin'].split(',') + x['kdr_S_origin'].split(','))\n",
    "    # There should be at least two 'wt' calls\n",
    "    if np.sum(joined_calls == 'wt') < 2:\n",
    "        print(f'Too many different mutant haplotype backgrounds in sample {sample_name}')\n",
    "        return('?,?')\n",
    "    # Otherwise, drop two wildtype calls\n",
    "    else:\n",
    "        which_drop = np.where(joined_calls == 'wt')[0][:2]\n",
    "        return(','.join(np.delete(joined_calls, which_drop)))\n",
    "\n",
    "def _get_single_gen_call_with_402(x): \n",
    "    if 'sample_name' in x.index:\n",
    "        sample_name = x['sample_name']\n",
    "    else:\n",
    "        sample_name = x.name\n",
    "    joined_calls = np.array(x['kdr_F_origin'].split(',') + \n",
    "                            x['kdr_S_origin'].split(',') +\n",
    "                            x['kdr_402_origin'].split(',')\n",
    "    )\n",
    "    # There should be at least four 'wt' calls\n",
    "    if np.sum(np.isin(joined_calls,  ['wt', '?'])) < 4:\n",
    "        print(f'Too many different mutant haplotype backgrounds in sample {sample_name}')\n",
    "        return('?,?')\n",
    "    # Otherwise, drop four wildtype calls\n",
    "    else:\n",
    "        which_drop = np.concatenate([\n",
    "            np.where(joined_calls == 'wt')[0],\n",
    "            np.where(joined_calls == '?')[0]\n",
    "        ])[:4]\n",
    "        return(','.join(np.delete(joined_calls, which_drop)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81488cf0",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Load metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fd1a611c",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(metadata_path, sep = '\\t', index_col = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ad48170",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Get the genotype calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c7dc9a80",
   "metadata": {
    "scrolled": true,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/eric/.local/lib/python3.8/site-packages/allel/io/vcf_read.py:1732: UserWarning: invalid INFO header: '##INFO=<ID=VDB,Number=1,Type=Float,Description=\"Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)\",Version=\"3\">\\n'\n",
      "  warnings.warn('invalid INFO header: %r' % header)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'samples': array(['GH_01', 'GH_02', 'GH_03', 'GH_04', 'GH_05', 'GH_06', 'GH_07',\n",
       "        'GH_08', 'GH_09', 'GH_10', 'GH_11', 'GH_12', 'GH_13', 'GH_14',\n",
       "        'GH_15', 'GH_16', 'GH_17', 'GH_18', 'GH_19', 'GH_20', 'GH_21',\n",
       "        'GH_22', 'GH_23', 'GH_24', 'GH_25', 'GH_26', 'GH_27', 'GH_28',\n",
       "        'GH_29', 'GH_30', 'GH_31', 'GH_32', 'GH_33', 'GH_34', 'GH_35',\n",
       "        'GH_36', 'GH_37', 'GH_38', 'GH_39', 'GH_40_low_vol', 'GH_41',\n",
       "        'GH_42', 'GH_43', 'GH_44', 'GH_45', 'GH_46', 'GH_47', 'GH_48',\n",
       "        'GH_49', 'GH_50', 'GH_51', 'GH_52', 'GH_53', 'GH_54', 'GH_55',\n",
       "        'GH_56', 'GH_57', 'GH_58', 'GH_59', 'GH_60', 'GH_61', 'GH_62',\n",
       "        'GH_63', 'GH_64', 'GH_65_low_vol', 'GH_66', 'GH_67', 'GH_68',\n",
       "        'GH_69', 'GH_70', 'GH_71', 'GH_72', 'GH_73', 'GH_74', 'GH_75',\n",
       "        'GH_76', 'GH_77', 'GH_78', 'GH_79', 'GH_80_low_vol', 'GH_81',\n",
       "        'GH_82', 'GH_83', 'GH_84', 'GH_85', 'GH_86', 'GH_87', 'GH_88',\n",
       "        'GH_89', 'GH_90', 'GH_91', 'GH_92', 'GH_93', 'GH_94', 'GH_95',\n",
       "        'GH_Negative', 'GM_01', 'GM_02', 'GM_03', 'GM_04', 'GM_05',\n",
       "        'GM_06', 'GM_07', 'GM_08', 'GM_09', 'GM_10', 'GM_11', 'Sante_10A',\n",
       "        'GM_13', 'GM_14', 'GM_15', 'GM_16', 'GM_17', 'GM_18', 'GM_19',\n",
       "        'GM_20', 'GM_21', 'GM_22', 'GM_23', 'Sante_10B', 'GM_25', 'GM_26',\n",
       "        'GM_27', 'GM_28', 'GM_29', 'GM_30', 'GM_31', 'GM_32', 'GM_33',\n",
       "        'GM_34', 'GM_35', 'Sante_10C', 'GM_37', 'GM_38', 'GM_39', 'GM_40',\n",
       "        'GM_41', 'GM_42', 'GM_43', 'GM_44', 'GM_45', 'GM_46', 'GM_47',\n",
       "        'Sante_10D', 'GM_49', 'GM_50', 'GM_51', 'GM_52', 'GM_53', 'GM_54',\n",
       "        'GM_55', 'GM_56', 'GM_57', 'GM_58', 'GM_59', 'Sante_10E', 'GM_61',\n",
       "        'GM_62', 'GM_63', 'GM_64', 'GM_65', 'GM_66', 'GM_67', 'GM_68',\n",
       "        'GM_69', 'GM_70', 'GM_71', 'Sante_10F', 'GM_73', 'GM_74', 'GM_75',\n",
       "        'GM_76', 'GM_77', 'GM_78', 'GM_79', 'GM_80', 'GM_81', 'GM_82',\n",
       "        'GM_83', 'Sante_10G', 'GM_85', 'GM_86', 'GM_87', 'GM_88', 'GM_89',\n",
       "        'GM_90', 'GM_91', 'GM_92', 'GM_93', 'GM_94', 'GM_95', 'GM_96',\n",
       "        'VK7_dead_01', 'VK7_dead_02', 'VK7_dead_03', 'VK7_dead_04',\n",
       "        'VK7_dead_05', 'VK7_dead_06', 'VK7_alive_01', 'VK7_alive_02',\n",
       "        'VK7_alive_03', 'VK7_alive_04', 'VK7_alive_05', 'VK7_alive_06',\n",
       "        'VK7_dead_07', 'VK7_dead_08', 'VK7_dead_09', 'VK7_dead_10',\n",
       "        'VK7_dead_11', 'VK7_dead_12', 'VK7_alive_07', 'VK7_alive_08',\n",
       "        'VK7_alive_09', 'VK7_alive_10', 'VK7_alive_11', 'VK7_alive_12',\n",
       "        'VK7_dead_13', 'VK7_dead_14', 'VK7_dead_15', 'VK7_dead_16',\n",
       "        'VK7_dead_17', 'VK7_dead_04_dil', 'VK7_alive_13', 'VK7_alive_14',\n",
       "        'VK7_alive_15', 'VK7_alive_16', 'VK7_alive_17', 'VK7_alive_18',\n",
       "        'VK7_dead_18', 'VK7_dead_19', 'VK7_dead_20', 'VK7_dead_21',\n",
       "        'VK7_dead_22', 'VK7_dead_10_dil', 'VK7_alive_19', 'VK7_alive_20',\n",
       "        'VK7_alive_21', 'VK7_alive_22', 'VK7_alive_23', 'VK7_alive_24',\n",
       "        'VK7_dead_23', 'VK7_dead_24', 'VK7_dead_25', 'VK7_dead_26',\n",
       "        'VK7_dead_27', 'VK7_dead_16_dil', 'VK7_alive_25', 'VK7_alive_26',\n",
       "        'VK7_alive_27', 'VK7_alive_28', 'VK7_alive_29', 'VK7_alive_30',\n",
       "        'VK7_dead_28', 'VK7_dead_29', 'VK7_dead_30', 'VK7_dead_31',\n",
       "        'VK7_dead_32', 'VK7_dead_22_dil', 'VK7_alive_31', 'VK7_alive_32',\n",
       "        'VK7_alive_33', 'VK7_alive_34', 'VK7_alive_35', 'VK7_alive_36',\n",
       "        'VK7_dead_33', 'VK7_dead_34', 'VK7_dead_35', 'VK7_dead_36',\n",
       "        'VK7_dead_37', 'VK7_dead_28_dil', 'VK7_alive_37', 'VK7_alive_38',\n",
       "        'VK7_alive_39', 'VK7_alive_40', 'VK7_alive_41', 'VK7_alive_42',\n",
       "        'VK7_dead_38', 'VK7_dead_39', 'VK7_dead_40', 'VK7_dead_41',\n",
       "        'VK7_dead_42', 'VK7_dead_34_dil', 'VK7_alive_43', 'VK7_alive_44',\n",
       "        'VK7_alive_45', 'VK7_alive_46', 'VK7_alive_47', 'VK7_Negative',\n",
       "        'Siaya_Delta_Dead_01', 'Siaya_Delta_Dead_02',\n",
       "        'Siaya_Delta_Dead_03', 'Siaya_Delta_Dead_04',\n",
       "        'Siaya_Delta_Dead_05', 'Siaya_Delta_Dead_06',\n",
       "        'Siaya_Delta_Dead_07', 'Siaya_Delta_Dead_08',\n",
       "        'Siaya_Delta_Dead_09', 'Siaya_Delta_Dead_10',\n",
       "        'Siaya_Delta_Dead_11', 'Siaya_Delta_Dead_12',\n",
       "        'Siaya_Delta_Dead_13', 'Siaya_Delta_Dead_14',\n",
       "        'Siaya_Delta_Dead_15', 'Siaya_Delta_Dead_16',\n",
       "        'Siaya_Delta_Dead_17', 'Siaya_Delta_Dead_18',\n",
       "        'Siaya_Delta_Dead_19', 'Siaya_Delta_Dead_20',\n",
       "        'Siaya_Delta_Dead_21', 'Siaya_Delta_Dead_22',\n",
       "        'Siaya_Delta_Dead_23', 'Siaya_Delta_Dead_24',\n",
       "        'Siaya_Delta_Dead_25', 'Siaya_Delta_Dead_26',\n",
       "        'Siaya_Delta_Dead_27', 'Siaya_Delta_Dead_28',\n",
       "        'Siaya_Delta_Dead_29', 'Siaya_Delta_Dead_30',\n",
       "        'Siaya_Delta_Dead_31', 'Siaya_Delta_Dead_32',\n",
       "        'Siaya_Delta_Dead_33', 'Siaya_Delta_Dead_34',\n",
       "        'Siaya_Delta_Dead_35', 'Siaya_Delta_Dead_36',\n",
       "        'Siaya_Delta_Dead_37', 'Siaya_Delta_Dead_38',\n",
       "        'Siaya_Delta_Dead_39', 'Siaya_Delta_Dead_40',\n",
       "        'Siaya_Delta_Dead_41', 'Siaya_Delta_Dead_42',\n",
       "        'Siaya_Delta_Dead_43', 'Siaya_Delta_Dead_44',\n",
       "        'Siaya_Delta_Dead_45', 'Siaya_Delta_Dead_46',\n",
       "        'Siaya_Delta_Dead_47', 'Siaya_Delta_Dead_48',\n",
       "        'Siaya_Delta_Dead_49', 'Siaya_Delta_Dead_50',\n",
       "        'Siaya_Delta_Dead_51', 'Siaya_Delta_Dead_52',\n",
       "        'Siaya_Delta_Dead_53', 'Siaya_Delta_Dead_54',\n",
       "        'Siaya_Delta_Dead_55', 'Siaya_Delta_Dead_56',\n",
       "        'Siaya_Delta_Dead_57', 'Siaya_Delta_Dead_58',\n",
       "        'Siaya_Delta_Dead_59', 'Siaya_Delta_Dead_60',\n",
       "        'Siaya_Delta_Dead_61', 'Siaya_Delta_Dead_62',\n",
       "        'Siaya_Delta_Dead_63', 'Siaya_Delta_Dead_64',\n",
       "        'Siaya_Delta_Dead_65', 'Siaya_Delta_Dead_66',\n",
       "        'Siaya_Delta_Dead_67', 'Siaya_Delta_Dead_68',\n",
       "        'Siaya_Delta_Dead_69', 'Siaya_Delta_Dead_70',\n",
       "        'Siaya_Delta_Dead_71', 'Siaya_Delta_Dead_72',\n",
       "        'Siaya_Delta_Dead_73', 'Siaya_Delta_Dead_74',\n",
       "        'Siaya_Delta_Dead_75', 'Siaya_Delta_Dead_76',\n",
       "        'Siaya_Delta_Dead_77', 'Siaya_Delta_Dead_78',\n",
       "        'Siaya_Delta_Dead_79', 'Siaya_Delta_Dead_80',\n",
       "        'Siaya_Delta_Dead_81', 'Siaya_Delta_Dead_82',\n",
       "        'Siaya_Delta_Dead_83', 'Siaya_Delta_Dead_84',\n",
       "        'Siaya_Delta_Dead_85', 'Siaya_Delta_Dead_86',\n",
       "        'Siaya_Delta_Dead_87', 'Siaya_Delta_Dead_88',\n",
       "        'Siaya_Delta_Dead_89', 'Siaya_Delta_Dead_90',\n",
       "        'Siaya_Delta_Dead_91', 'Siaya_Delta_Dead_92',\n",
       "        'Siaya_Delta_Dead_93', 'Siaya_Delta_Dead_94',\n",
       "        'Siaya_Delta_Dead_95', 'Siaya_Delta_Dead_Negative', 'Calvin_01',\n",
       "        'Calvin_02', 'Calvin_03', 'Calvin_04', 'Calvin_05', 'Calvin_06',\n",
       "        'Calvin_07', 'Calvin_08', 'Calvin_09', 'Calvin_10', 'Calvin_11',\n",
       "        'Calvin_12', 'Calvin_13', 'Calvin_14', 'Calvin_15', 'Calvin_16',\n",
       "        'Calvin_17', 'Calvin_18', 'Calvin_19', 'Calvin_20', 'Calvin_21',\n",
       "        'Calvin_22', 'Calvin_23', 'Calvin_24', 'Calvin_25', 'Calvin_26',\n",
       "        'Calvin_27', 'Calvin_28', 'Calvin_29', 'Calvin_30', 'Calvin_31',\n",
       "        'Calvin_32', 'Calvin_33', 'Calvin_34', 'Calvin_35', 'Calvin_36',\n",
       "        'Calvin_37', 'Calvin_38', 'Calvin_39', 'Calvin_40', 'Calvin_41',\n",
       "        'Calvin_42', 'Calvin_43', 'Calvin_44', 'Calvin_45', 'Calvin_46',\n",
       "        'Calvin_47', 'Calvin_48', 'Calvin_49', 'Calvin_50', 'Calvin_51',\n",
       "        'Calvin_52', 'Calvin_53', 'Calvin_54', 'Calvin_55', 'Calvin_56',\n",
       "        'Calvin_57', 'Calvin_58', 'Calvin_59', 'Calvin_60', 'Calvin_61',\n",
       "        'Calvin_62', 'Calvin_63', 'Calvin_64', 'Calvin_65', 'Calvin_66',\n",
       "        'Calvin_67', 'Calvin_68', 'Calvin_69', 'Calvin_70', 'Calvin_71',\n",
       "        'Calvin_72', 'Calvin_73', 'Calvin_74', 'Calvin_75', 'Calvin_76',\n",
       "        'Calvin_77', 'Calvin_78', 'Calvin_79', 'Calvin_80', 'Calvin_81',\n",
       "        'Calvin_82', 'Calvin_83', 'Calvin_84', 'Calvin_85', 'Calvin_86',\n",
       "        'Calvin_87', 'Calvin_88', 'Calvin_89', 'Calvin_90', 'Calvin_91',\n",
       "        'Calvin_92', 'Calvin_93', 'Calvin_94', 'Calvin_95',\n",
       "        'Calvin_Negative', 'Siaya_Delta_Alive_01', 'Siaya_Delta_Alive_02',\n",
       "        'Siaya_Delta_Alive_03', 'Siaya_Delta_Alive_04',\n",
       "        'Siaya_Delta_Alive_05', 'Siaya_Delta_Alive_06',\n",
       "        'Siaya_Delta_Alive_07', 'Siaya_Delta_Alive_08',\n",
       "        'Siaya_Delta_Alive_09', 'Siaya_Delta_Alive_10',\n",
       "        'Siaya_Delta_Alive_11', 'Siaya_Delta_Alive_12',\n",
       "        'Siaya_Delta_Alive_13', 'Siaya_Delta_Alive_14',\n",
       "        'Siaya_Delta_Alive_15', 'Siaya_Delta_Alive_16',\n",
       "        'Siaya_Delta_Alive_17', 'Siaya_Delta_Alive_18',\n",
       "        'Siaya_Delta_Alive_19', 'Siaya_Delta_Alive_20',\n",
       "        'Siaya_Delta_Alive_21', 'Siaya_Delta_Alive_22',\n",
       "        'Siaya_Delta_Alive_23', 'Siaya_Delta_Alive_24',\n",
       "        'Siaya_Delta_Alive_25', 'Siaya_Delta_Alive_26',\n",
       "        'Siaya_Delta_Alive_27', 'Siaya_Delta_Alive_28',\n",
       "        'Siaya_Delta_Alive_29', 'Siaya_Delta_Alive_30',\n",
       "        'Siaya_Delta_Alive_31', 'Siaya_Delta_Alive_32',\n",
       "        'Siaya_Delta_Alive_33', 'Siaya_Delta_Alive_34',\n",
       "        'Siaya_Delta_Alive_35', 'Siaya_Delta_Alive_36',\n",
       "        'Siaya_Delta_Alive_37', 'Siaya_Delta_Alive_38',\n",
       "        'Siaya_Delta_Alive_39', 'Siaya_Delta_Alive_40',\n",
       "        'Siaya_Delta_Alive_41', 'Siaya_Delta_Alive_42',\n",
       "        'Siaya_Delta_Alive_43', 'Siaya_Delta_Alive_44',\n",
       "        'Siaya_Delta_Alive_45', 'Siaya_Delta_Alive_46',\n",
       "        'Siaya_Delta_Alive_47', 'Siaya_Delta_Alive_48',\n",
       "        'Siaya_Delta_Alive_49', 'Siaya_Delta_Alive_50',\n",
       "        'Siaya_Delta_Alive_51', 'Siaya_Delta_Alive_52',\n",
       "        'Siaya_Delta_Alive_53', 'Siaya_Delta_Alive_54',\n",
       "        'Siaya_Delta_Alive_55', 'Siaya_Delta_Alive_56',\n",
       "        'Siaya_Delta_Alive_57', 'Siaya_Delta_Alive_58',\n",
       "        'Siaya_Delta_Alive_59', 'Siaya_Delta_Alive_60',\n",
       "        'Siaya_Delta_Alive_61', 'Siaya_Delta_Alive_62',\n",
       "        'Siaya_Delta_Alive_63', 'Siaya_Delta_Alive_64',\n",
       "        'Siaya_Delta_Alive_65', 'Siaya_Delta_Alive_66',\n",
       "        'Siaya_Delta_Alive_67', 'Siaya_Delta_Alive_68',\n",
       "        'Siaya_Delta_Alive_69', 'Siaya_Delta_Alive_70',\n",
       "        'Siaya_Delta_Alive_71', 'Siaya_Delta_Alive_72',\n",
       "        'Siaya_Delta_Alive_73', 'Siaya_Delta_Alive_74',\n",
       "        'Siaya_Delta_Alive_75', 'Siaya_Delta_Alive_76',\n",
       "        'Siaya_Delta_Alive_77', 'Siaya_Delta_Alive_78',\n",
       "        'Siaya_Delta_Alive_79', 'Siaya_Delta_Alive_80',\n",
       "        'Siaya_Delta_Alive_81', 'Siaya_Delta_Alive_82',\n",
       "        'Siaya_Delta_Alive_83', 'Siaya_Delta_Alive_84',\n",
       "        'Siaya_Delta_Alive_85', 'Siaya_Delta_Alive_86',\n",
       "        'Siaya_Delta_Alive_87', 'Siaya_Delta_Alive_88',\n",
       "        'Siaya_Delta_Alive_89', 'Siaya_Delta_Alive_90',\n",
       "        'Siaya_Delta_Alive_91', 'Siaya_Delta_Alive_92',\n",
       "        'Siaya_Delta_Alive_93', 'Siaya_Delta_Alive_94',\n",
       "        'Siaya_Delta_Alive_95', 'Siaya_Delta_Alive_Negative',\n",
       "        'Siaya_Delta_Dead_96', 'Siaya_Delta_Dead_97',\n",
       "        'Siaya_Delta_Dead_98', 'Siaya_Delta_Dead_99',\n",
       "        'Siaya_Delta_Dead_100', 'Siaya_Delta_Alive_96',\n",
       "        'Siaya_Delta_Alive_97', 'Siaya_Delta_Alive_98',\n",
       "        'Siaya_Delta_Alive_99', 'ML_01', 'ML_02', 'ML_03',\n",
       "        'Siaya_Delta_Dead_101', 'Siaya_Delta_Dead_102',\n",
       "        'Siaya_Delta_Dead_103', 'Siaya_Delta_Dead_104',\n",
       "        'Siaya_Delta_Dead_105', 'Siaya_Delta_Alive_100',\n",
       "        'Siaya_Delta_Alive_101', 'Siaya_Delta_Alive_102',\n",
       "        'Siaya_Delta_Alive_103', 'ML_04', 'ML_05', 'ML_06',\n",
       "        'Siaya_Delta_Dead_106', 'Siaya_Delta_Dead_107',\n",
       "        'Siaya_Delta_Dead_108', 'Siaya_Delta_Dead_109',\n",
       "        'Siaya_Delta_Dead_110', 'Siaya_Delta_Alive_104',\n",
       "        'Siaya_Delta_Alive_105', 'Siaya_Delta_Alive_106',\n",
       "        'Siaya_Delta_Alive_107', 'ML_07', 'ML_08', 'random_arabiensis',\n",
       "        'Siaya_Delta_Dead_111', 'Siaya_Delta_Dead_112',\n",
       "        'Siaya_Delta_Dead_113', 'Siaya_Delta_Dead_114',\n",
       "        'Siaya_Delta_Dead_115', 'Siaya_Delta_Alive_108',\n",
       "        'Siaya_Delta_Alive_109', 'Siaya_Delta_Alive_110',\n",
       "        'Siaya_Delta_Alive_111', 'ML_09', 'ML_10', 'random_coluzzii',\n",
       "        'Siaya_Delta_Dead_116', 'Siaya_Delta_Dead_117',\n",
       "        'Siaya_Delta_Dead_118', 'Siaya_Delta_Dead_119',\n",
       "        'Siaya_Delta_Alive_112', 'Siaya_Delta_Alive_113',\n",
       "        'Siaya_Delta_Alive_114', 'Siaya_Delta_Alive_115',\n",
       "        'Siaya_Delta_Alive_116', 'ML_11', 'ML_12', 'random_funestus',\n",
       "        'Siaya_Delta_Dead_120', 'Siaya_Delta_Dead_121',\n",
       "        'Siaya_Delta_Dead_122', 'Siaya_Delta_Dead_123',\n",
       "        'Siaya_Delta_Alive_117', 'Siaya_Delta_Alive_118',\n",
       "        'Siaya_Delta_Alive_119', 'Siaya_Delta_Alive_120',\n",
       "        'Siaya_Delta_Alive_121', 'ML_13', 'ML_14', 'random_stephensi',\n",
       "        'Siaya_Delta_Dead_124', 'Siaya_Delta_Dead_125',\n",
       "        'Siaya_Delta_Dead_126', 'Siaya_Delta_Dead_127',\n",
       "        'Siaya_Delta_Alive_122', 'Siaya_Delta_Alive_123',\n",
       "        'Siaya_Delta_Alive_124', 'Siaya_Delta_Alive_125',\n",
       "        'Siaya_Delta_Alive_126', 'ML_15', 'ML_16', 'random_stephensi2',\n",
       "        'Siaya_Delta_Dead_128', 'Siaya_Delta_Dead_129',\n",
       "        'Siaya_Delta_Dead_130', 'Siaya_Delta_Dead_131',\n",
       "        'Siaya_Delta_Alive_127', 'Siaya_Delta_Alive_128',\n",
       "        'Siaya_Delta_Alive_129', 'Siaya_Delta_Alive_130',\n",
       "        'Siaya_Delta_Alive_131', 'ML_17', 'ML_18', 'mixedplate_negative'],\n",
       "       dtype=object),\n",
       " 'calldata/GT': array([[[ 1,  1],\n",
       "         [ 1,  1],\n",
       "         [ 1,  1],\n",
       "         ...,\n",
       "         [ 1,  1],\n",
       "         [-1, -1],\n",
       "         [-1, -1]],\n",
       " \n",
       "        [[ 1,  1],\n",
       "         [ 1,  1],\n",
       "         [-1, -1],\n",
       "         ...,\n",
       "         [-1, -1],\n",
       "         [ 0,  0],\n",
       "         [-1, -1]],\n",
       " \n",
       "        [[ 1,  1],\n",
       "         [ 1,  1],\n",
       "         [-1, -1],\n",
       "         ...,\n",
       "         [-1, -1],\n",
       "         [-1, -1],\n",
       "         [-1, -1]],\n",
       " \n",
       "        ...,\n",
       " \n",
       "        [[ 1,  1],\n",
       "         [ 1,  1],\n",
       "         [-1, -1],\n",
       "         ...,\n",
       "         [ 1,  1],\n",
       "         [ 0,  1],\n",
       "         [-1, -1]],\n",
       " \n",
       "        [[ 1,  1],\n",
       "         [ 1,  1],\n",
       "         [ 0,  0],\n",
       "         ...,\n",
       "         [-1, -1],\n",
       "         [ 0,  0],\n",
       "         [ 0,  0]],\n",
       " \n",
       "        [[ 1,  1],\n",
       "         [ 1,  1],\n",
       "         [-1, -1],\n",
       "         ...,\n",
       "         [ 1,  1],\n",
       "         [-1, -1],\n",
       "         [-1, -1]]], dtype=int8),\n",
       " 'variants/ALT': array([['A', 'T', ''],\n",
       "        ['C', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['A', 'C', ''],\n",
       "        ['', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['C', 'A', ''],\n",
       "        ['G', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['AG', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['G', 'A', ''],\n",
       "        ['G', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['', '', ''],\n",
       "        ['', '', ''],\n",
       "        ['', '', ''],\n",
       "        ['', '', ''],\n",
       "        ['', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['T', 'G', ''],\n",
       "        ['A', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['T', 'G', ''],\n",
       "        ['G', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['A', 'T', ''],\n",
       "        ['G', 'C', 'A'],\n",
       "        ['A', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['T', 'A', ''],\n",
       "        ['G', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['A', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['T', 'G', ''],\n",
       "        ['T', '', ''],\n",
       "        ['', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['C', 'T', ''],\n",
       "        ['T', '', ''],\n",
       "        ['T', 'C', ''],\n",
       "        ['G', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ['T', '', '']], dtype=object),\n",
       " 'variants/CHROM': array(['2R', '2R', '2R', '2R', '2R', '2R', '2R', '2R', '2R', '2R', '2R',\n",
       "        '2R', '2R', '2R', '2R', '2R', '2R', '3R', '3R', '3R', '3R', '3R',\n",
       "        '3R', '3R', '3R', '3R', '3R', '3R', '3R', '3R', '2L', '2L', '2L',\n",
       "        '2L', '2L', '2L', '2L', '2L', '2L', '2L', '2L', '2L', '2L', '2L',\n",
       "        '2L', '2L', '2L', '2L', '2L', '2L', '2L', '2L', '2L', '2L', '2L',\n",
       "        '2L', '2L', '2L', '2L', '2L', '2L', '2L', '2L', '2L', '2L', '2L',\n",
       "        '2L', '2L', '2L', '2L', '2L', '2L', '2L', '3L', '3L', '3L', '3L',\n",
       "        '3L', '3L', '3L', '3L', '3L', 'X', 'X', 'X', 'X', 'X', 'X', 'X',\n",
       "        'X', 'X', 'X', 'X', 'X', 'X', 'X'], dtype=object),\n",
       " 'variants/FILTER_PASS': array([False, False, False, False, False, False, False, False, False,\n",
       "        False, False, False, False, False, False, False, False, False,\n",
       "        False, False, False, False, False, False, False, False, False,\n",
       "        False, False, False, False, False, False, False, False, False,\n",
       "        False, False, False, False, False, False, False, False, False,\n",
       "        False, False, False, False, False, False, False, False, False,\n",
       "        False, False, False, False, False, False, False, False, False,\n",
       "        False, False, False, False, False, False, False, False, False,\n",
       "        False, False, False, False, False, False, False, False, False,\n",
       "        False, False, False, False, False, False, False, False, False,\n",
       "        False, False, False, False, False, False]),\n",
       " 'variants/ID': array(['.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.',\n",
       "        '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.',\n",
       "        '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.',\n",
       "        '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.',\n",
       "        '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.',\n",
       "        '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.',\n",
       "        '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.',\n",
       "        '.', '.', '.', '.', '.'], dtype=object),\n",
       " 'variants/POS': array([ 3492074,  8368731, 17854287, 28492879, 28494247, 28497563,\n",
       "        28497967, 28498192, 28499661, 28500643, 28501399, 28501399,\n",
       "        28502736, 28502850, 48714551, 48714601, 49438586,    34264,\n",
       "           35121,    38161,    42848,    50590,    71973,    97617,\n",
       "         8564156, 28592158, 28595449, 28598062, 28598166, 28600701,\n",
       "          209536,   927247,  1274353,  1274353,  1418210,  1571929,\n",
       "         1776348,  1947574,  2380982,  2381706,  2388595,  2391229,\n",
       "         2391229,  2399997,  2403941,  2408423,  2408677,  2414062,\n",
       "         2416980,  2422651,  2422652,  2425052,  2425766,  2429618,\n",
       "         2429745,  2429897,  2430424,  2430817,  2430863,  2430880,\n",
       "         2430881,  2431005,  2431061,  2432419,  2435581, 20288132,\n",
       "        25429235, 25429236, 25635973, 34073195, 34100758, 34118141,\n",
       "        34118260,    65133,   129051,   193173,   367248,   380974,\n",
       "          413508, 11208465, 11213955, 11213955, 15249654, 15251490,\n",
       "        15255321, 18758300, 19631625, 20014696, 20128328, 20129288,\n",
       "        21721801, 22164043, 22798648, 23468268, 24229846, 24266728],\n",
       "       dtype=int32),\n",
       " 'variants/QUAL': array([284.59  , 284.59  , 284.59  , 284.59  , 284.59  , 284.59  ,\n",
       "        284.59  , 145.588 , 284.59  , 284.59  , 284.59  ,  26.4692,\n",
       "        284.59  , 284.59  , 284.59  , 284.59  , 284.59  , 284.59  ,\n",
       "        284.59  , 284.59  , 284.59  , 284.59  , 284.59  , 284.59  ,\n",
       "        284.59  , 284.59  , 284.59  , 284.59  , 284.59  , 284.59  ,\n",
       "        284.59  , 284.59  , 284.59  ,  58.36  , 284.59  , 284.59  ,\n",
       "        284.59  , 241.589 , 275.59  , 284.59  , 284.59  , 284.59  ,\n",
       "         22.5809, 284.59  , 284.59  , 284.59  , 284.59  , 284.59  ,\n",
       "        284.59  , 284.59  , 284.59  , 284.59  , 284.59  , 284.59  ,\n",
       "        272.59  , 284.59  , 284.59  , 284.59  , 284.59  , 284.59  ,\n",
       "        284.59  , 271.59  , 284.59  , 284.59  , 284.59  , 284.59  ,\n",
       "        284.59  , 284.59  , 284.59  , 284.59  , 284.59  , 284.59  ,\n",
       "        284.59  , 228.416 , 284.59  , 284.59  , 284.59  , 284.59  ,\n",
       "        284.59  , 284.59  , 284.59  ,  30.2881, 284.59  , 284.59  ,\n",
       "        268.589 , 284.59  , 284.59  , 284.59  , 284.59  , 284.59  ,\n",
       "        284.59  , 284.59  , 284.59  , 284.59  , 284.59  , 284.59  ],\n",
       "       dtype=float32),\n",
       " 'variants/REF': array(['G', 'A', 'C', 'A', 'C', 'T', 'G', 'C', 'G', 'G', 'G', 'G', 'C',\n",
       "        'C', 'A', 'A', 'T', 'A', 'A', 'A', 'A', 'T', 'C', 'G', 'G', 'G',\n",
       "        'C', 'G', 'A', 'A', 'G', 'A', 'A', 'A', 'C', 'C', 'C', 'A', 'T',\n",
       "        'A', 'G', 'T', 'T', 'G', 'T', 'C', 'A', 'A', 'C', 'T', 'A', 'G',\n",
       "        'T', 'T', 'A', 'A', 'G', 'G', 'T', 'C', 'C', 'C', 'C', 'T', 'G',\n",
       "        'T', 'G', 'C', 'G', 'C', 'A', 'G', 'G', 'G', 'A', 'A', 'C', 'C',\n",
       "        'A', 'A', 'G', 'G', 'A', 'C', 'A', 'C', 'C', 'A', 'C', 'G', 'C',\n",
       "        'C', 'A', 'A', 'A', 'A'], dtype=object)}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_snp_calls_dict = allel.read_vcf(vcf_path)\n",
    "vcf_snp_calls_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "507e6744",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Filter the SNPs to just the ones useful for kdr origin analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a5de2ea6",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[[ 0,  0],\n",
       "        [ 0,  0],\n",
       "        [-1, -1],\n",
       "        ...,\n",
       "        [ 0,  0],\n",
       "        [-1, -1],\n",
       "        [-1, -1]],\n",
       "\n",
       "       [[ 0,  0],\n",
       "        [ 0,  0],\n",
       "        [-1, -1],\n",
       "        ...,\n",
       "        [ 0,  0],\n",
       "        [-1, -1],\n",
       "        [-1, -1]],\n",
       "\n",
       "       [[ 0,  0],\n",
       "        [ 0,  0],\n",
       "        [-1, -1],\n",
       "        ...,\n",
       "        [ 0,  0],\n",
       "        [-1, -1],\n",
       "        [-1, -1]],\n",
       "\n",
       "       ...,\n",
       "\n",
       "       [[ 0,  0],\n",
       "        [ 0,  0],\n",
       "        [ 0,  0],\n",
       "        ...,\n",
       "        [ 0,  0],\n",
       "        [ 0,  0],\n",
       "        [-1, -1]],\n",
       "\n",
       "       [[ 0,  0],\n",
       "        [ 0,  0],\n",
       "        [ 0,  0],\n",
       "        ...,\n",
       "        [ 1,  1],\n",
       "        [ 0,  0],\n",
       "        [ 0,  0]],\n",
       "\n",
       "       [[ 1,  1],\n",
       "        [ 1,  1],\n",
       "        [ 0,  1],\n",
       "        ...,\n",
       "        [ 0,  0],\n",
       "        [ 0,  1],\n",
       "        [-1, -1]]], dtype=int8)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "which_snps = (vcf_snp_calls_dict['variants/CHROM'] == '2L') & np.isin(vcf_snp_calls_dict['variants/POS'], hap_def['variant_pos']) \n",
    "snp_calls = vcf_snp_calls_dict['calldata/GT'][which_snps, :, :]\n",
    "pos = vcf_snp_calls_dict['variants/POS'][which_snps]\n",
    "alt = vcf_snp_calls_dict['variants/ALT'][which_snps, :]\n",
    "ref = vcf_snp_calls_dict['variants/REF'][which_snps]\n",
    "snp_calls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2710ba21",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Convert genotype calls to nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5f235e86",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>SNP name</th>\n",
       "      <th>Def-S1-3</th>\n",
       "      <th>Def-F5-2</th>\n",
       "      <th>Def-S4S5-2</th>\n",
       "      <th>Def-F3</th>\n",
       "      <th>Def-S4S5</th>\n",
       "      <th>Def-S2S4</th>\n",
       "      <th>Def-S5</th>\n",
       "      <th>Def-F3F4-2</th>\n",
       "      <th>kdr-995S</th>\n",
       "      <th>kdr-995F</th>\n",
       "      <th>Def-F2</th>\n",
       "      <th>Def-S2-4</th>\n",
       "      <th>Def-S3</th>\n",
       "      <th>Def-F1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GH_01</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_02</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_03</th>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>TT</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_04</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_05</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_130</th>\n",
       "      <td>CT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>CT</td>\n",
       "      <td>AT</td>\n",
       "      <td>AG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_131</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ML_17</th>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>??</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ML_18</th>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>??</td>\n",
       "      <td>AA</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mixedplate_negative</th>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>TT</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>AA</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>??</td>\n",
       "      <td>TT</td>\n",
       "      <td>??</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>672 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "SNP name              Def-S1-3 Def-F5-2 Def-S4S5-2 Def-F3 Def-S4S5 Def-S2S4  \\\n",
       "GH_01                       TT       AA         GG     GG       TT       CC   \n",
       "GH_02                       TT       AA         GG     GG       TT       CC   \n",
       "GH_03                       ??       ??         ??     ??       TT       ??   \n",
       "GH_04                       TT       AA         GG     GG       TT       CC   \n",
       "GH_05                       TT       AA         GG     GG       TT       CC   \n",
       "...                        ...      ...        ...    ...      ...      ...   \n",
       "Siaya_Delta_Alive_130       CT       AA         GG     GG       TT       CC   \n",
       "Siaya_Delta_Alive_131       TT       AA         GG     GG       TT       CC   \n",
       "ML_17                       TT       AA         GG     ??       TT       CC   \n",
       "ML_18                       ??       ??         ??     GG       TT       CC   \n",
       "mixedplate_negative         ??       ??         ??     ??       TT       ??   \n",
       "\n",
       "SNP name              Def-S5 Def-F3F4-2 kdr-995S kdr-995F Def-F2 Def-S2-4  \\\n",
       "GH_01                     AA         AA       TT       TT     GG       TT   \n",
       "GH_02                     AA         AA       TT       TT     GG       TT   \n",
       "GH_03                     ??         ??       ??       ??     ??       TT   \n",
       "GH_04                     AA         AA       TT       TT     GG       TT   \n",
       "GH_05                     AA         AA       TT       TT     GG       TT   \n",
       "...                      ...        ...      ...      ...    ...      ...   \n",
       "Siaya_Delta_Alive_130     AA         AA       CT       AT     AG       TT   \n",
       "Siaya_Delta_Alive_131     AA         AA       TT       TT     AA       TT   \n",
       "ML_17                     AA         AA       CC       AA     AA       TT   \n",
       "ML_18                     ??         AA       ??       ??     ??       TT   \n",
       "mixedplate_negative       ??         AA       ??       ??     ??       ??   \n",
       "\n",
       "SNP name              Def-S3 Def-F1  \n",
       "GH_01                     TT     AA  \n",
       "GH_02                     TT     AA  \n",
       "GH_03                     TT     AG  \n",
       "GH_04                     TT     AA  \n",
       "GH_05                     TT     AA  \n",
       "...                      ...    ...  \n",
       "Siaya_Delta_Alive_130     TT     GG  \n",
       "Siaya_Delta_Alive_131     TT     GG  \n",
       "ML_17                     GG     GG  \n",
       "ML_18                     TT     AG  \n",
       "mixedplate_negative       TT     ??  \n",
       "\n",
       "[672 rows x 14 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine ref and alt into a single matrix, and add a column of '?' at the end, so that \n",
    "# any genotype call of -1 (missing) draws the '?' character'\n",
    "snp_alleles = np.concatenate([np.reshape(ref, (len(ref), 1)), \n",
    "                              alt,\n",
    "                              np.full((len(ref), 1), '?')], \n",
    "                             axis = 1)\n",
    "\n",
    "# Convert numberic calls to nucleotides, and sort each pair of nucleotides alphabetically\n",
    "# (so, eg, the genotype 'TA' becomes 'AT')\n",
    "snp_genotypes_3d = snp_alleles[\n",
    "    np.array(range(snp_alleles.shape[0])).reshape(snp_alleles.shape[0], 1, 1), \n",
    "    snp_calls\n",
    "]\n",
    "snp_genotypes = np.apply_along_axis(lambda x: ''.join(np.sort(x)), 2, snp_genotypes_3d)\n",
    "\n",
    "# Store results in data frame\n",
    "hap_def.index = hap_def['variant_pos']\n",
    "gen_df = pd.DataFrame(\n",
    "    np.transpose(snp_genotypes), \n",
    "    index = vcf_snp_calls_dict['samples'],\n",
    "    columns = hap_def.loc[pos, 'SNP name']\n",
    ")\n",
    "gen_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb8620cc",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Obtain kdr origin calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b8eb15c6",
   "metadata": {
    "scrolled": true,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unexpected kdr F genotype. GH_03 ??\n",
      "Unexpected kdr S genotype. GH_03 ??\n",
      "Unexpected kdr F genotype. GH_08 ??\n",
      "Unexpected kdr S genotype. GH_08 ??\n",
      "Unexpected kdr F genotype. GH_09 ??\n",
      "Unexpected kdr S genotype. GH_09 ??\n",
      "Unexpected kdr F genotype. GH_24 ??\n",
      "Unexpected kdr S genotype. GH_24 ??\n",
      "Unexpected kdr F genotype. GH_36 ??\n",
      "Unexpected kdr S genotype. GH_36 ??\n",
      "Unexpected kdr F genotype. GH_48 ??\n",
      "Unexpected kdr S genotype. GH_48 ??\n",
      "Unexpected kdr F genotype. GH_60 ??\n",
      "Unexpected kdr S genotype. GH_60 ??\n",
      "Unexpected kdr F genotype. GH_84 ??\n",
      "Unexpected kdr S genotype. GH_84 ??\n",
      "Unexpected kdr F genotype. GH_89 ??\n",
      "Unexpected kdr S genotype. GH_89 ??\n",
      "Unexpected kdr F genotype. GH_92 ??\n",
      "Unexpected kdr S genotype. GH_92 ??\n",
      "Unexpected kdr F genotype. GH_Negative ??\n",
      "Unexpected kdr S genotype. GH_Negative ??\n",
      "Unexpected kdr F genotype. GM_08 ??\n",
      "Unexpected kdr S genotype. GM_08 ??\n",
      "Unexpected kdr F genotype. GM_21 ??\n",
      "Unexpected kdr S genotype. GM_21 ??\n",
      "Unexpected kdr F genotype. GM_27 ??\n",
      "Unexpected kdr S genotype. GM_27 ??\n",
      "Unexpected kdr F genotype. GM_28 ??\n",
      "Unexpected kdr S genotype. GM_28 ??\n",
      "Unexpected kdr F genotype. GM_42 ??\n",
      "Unexpected kdr S genotype. GM_42 ??\n",
      "Unexpected kdr F genotype. GM_49 ??\n",
      "Unexpected kdr S genotype. GM_49 ??\n",
      "Unexpected kdr F genotype. GM_56 ??\n",
      "Unexpected kdr S genotype. GM_56 ??\n",
      "Unexpected kdr F genotype. Sante_10G ??\n",
      "Unexpected kdr S genotype. Sante_10G ??\n",
      "Unexpected kdr F genotype. GM_93 ??\n",
      "Unexpected kdr S genotype. GM_93 ??\n",
      "Unexpected kdr F genotype. GM_96 ??\n",
      "Unexpected kdr S genotype. GM_96 ??\n",
      "Unexpected kdr F genotype. VK7_alive_37 ??\n",
      "Unexpected kdr S genotype. VK7_alive_37 ??\n",
      "Unexpected kdr F genotype. VK7_Negative ??\n",
      "Unexpected kdr S genotype. VK7_Negative ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_08 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_08 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_11 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_11 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_16 GG\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_24 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_24 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_37 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_37 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_48 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_48 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_49 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_49 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_52 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_52 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_72 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_72 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_74 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_74 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_76 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_76 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_83 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_83 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_90 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_90 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_91 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_91 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Dead_Negative ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Dead_Negative ??\n",
      "Unexpected kdr F genotype. Calvin_63 ??\n",
      "Unexpected kdr S genotype. Calvin_63 ??\n",
      "Unexpected kdr F genotype. Calvin_Negative ??\n",
      "Unexpected kdr S genotype. Calvin_Negative ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Alive_03 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Alive_03 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Alive_07 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Alive_07 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Alive_15 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Alive_15 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Alive_48 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Alive_48 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Alive_60 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Alive_60 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Alive_81 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Alive_81 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Alive_91 ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Alive_91 ??\n",
      "Unexpected kdr F genotype. Siaya_Delta_Alive_Negative ??\n",
      "Unexpected kdr S genotype. Siaya_Delta_Alive_Negative ??\n",
      "Unexpected kdr F genotype. ML_03 ??\n",
      "Unexpected kdr S genotype. ML_03 ??\n",
      "Unexpected kdr F genotype. ML_04 ??\n",
      "Unexpected kdr S genotype. ML_04 ??\n",
      "Unexpected kdr F genotype. ML_13 ??\n",
      "Unexpected kdr S genotype. ML_13 ??\n",
      "Unexpected kdr F genotype. random_stephensi ??\n",
      "Unexpected kdr S genotype. random_stephensi ??\n",
      "Unexpected kdr F genotype. random_stephensi2 ??\n",
      "Unexpected kdr S genotype. random_stephensi2 ??\n",
      "Unexpected kdr F genotype. ML_18 ??\n",
      "Unexpected kdr S genotype. ML_18 ??\n",
      "Unexpected kdr F genotype. mixedplate_negative ??\n",
      "Unexpected kdr S genotype. mixedplate_negative ??\n",
      "Too many different mutant haplotype backgrounds in sample GH_03\n",
      "Too many different mutant haplotype backgrounds in sample GH_08\n",
      "Too many different mutant haplotype backgrounds in sample GH_09\n",
      "Too many different mutant haplotype backgrounds in sample GH_24\n",
      "Too many different mutant haplotype backgrounds in sample GH_36\n",
      "Too many different mutant haplotype backgrounds in sample GH_48\n",
      "Too many different mutant haplotype backgrounds in sample GH_60\n",
      "Too many different mutant haplotype backgrounds in sample GH_84\n",
      "Too many different mutant haplotype backgrounds in sample GH_89\n",
      "Too many different mutant haplotype backgrounds in sample GH_92\n",
      "Too many different mutant haplotype backgrounds in sample GH_Negative\n",
      "Too many different mutant haplotype backgrounds in sample GM_08\n",
      "Too many different mutant haplotype backgrounds in sample GM_21\n",
      "Too many different mutant haplotype backgrounds in sample GM_27\n",
      "Too many different mutant haplotype backgrounds in sample GM_28\n",
      "Too many different mutant haplotype backgrounds in sample GM_42\n",
      "Too many different mutant haplotype backgrounds in sample GM_49\n",
      "Too many different mutant haplotype backgrounds in sample GM_56\n",
      "Too many different mutant haplotype backgrounds in sample Sante_10G\n",
      "Too many different mutant haplotype backgrounds in sample GM_93\n",
      "Too many different mutant haplotype backgrounds in sample GM_96\n",
      "Too many different mutant haplotype backgrounds in sample VK7_alive_37\n",
      "Too many different mutant haplotype backgrounds in sample VK7_Negative\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_03\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_08\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_11\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_16\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_24\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_37\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_38\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_39\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_47\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_48\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_49\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_52\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_58\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_70\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_71\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_72\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_73\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_74\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_76\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_83\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_90\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_91\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_Negative\n",
      "Too many different mutant haplotype backgrounds in sample Calvin_63\n",
      "Too many different mutant haplotype backgrounds in sample Calvin_Negative\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_03\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_07\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_15\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_33\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_48\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_60\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_66\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_69\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_80\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_81\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_85\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_87\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_89\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_91\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_Negative\n",
      "Too many different mutant haplotype backgrounds in sample ML_03\n",
      "Too many different mutant haplotype backgrounds in sample ML_04\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_108\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_114\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_109\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Dead_119\n",
      "Too many different mutant haplotype backgrounds in sample ML_13\n",
      "Too many different mutant haplotype backgrounds in sample random_stephensi\n",
      "Too many different mutant haplotype backgrounds in sample random_stephensi2\n",
      "Too many different mutant haplotype backgrounds in sample Siaya_Delta_Alive_128\n",
      "Too many different mutant haplotype backgrounds in sample ML_18\n",
      "Too many different mutant haplotype backgrounds in sample mixedplate_negative\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kdr_F_origin</th>\n",
       "      <th>kdr_S_origin</th>\n",
       "      <th>kdr_origin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GH_01</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_02</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_03</th>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_04</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_05</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_130</th>\n",
       "      <td>wt,F2</td>\n",
       "      <td>wt,S1</td>\n",
       "      <td>F2,S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_131</th>\n",
       "      <td>F2,F2</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F2,F2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ML_17</th>\n",
       "      <td>wt,wt</td>\n",
       "      <td>S3,S3</td>\n",
       "      <td>S3,S3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ML_18</th>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mixedplate_negative</th>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>672 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      kdr_F_origin kdr_S_origin kdr_origin\n",
       "GH_01                        F1,F1        wt,wt      F1,F1\n",
       "GH_02                        F1,F1        wt,wt      F1,F1\n",
       "GH_03                          ?,?          ?,?        ?,?\n",
       "GH_04                        F1,F1        wt,wt      F1,F1\n",
       "GH_05                        F1,F1        wt,wt      F1,F1\n",
       "...                            ...          ...        ...\n",
       "Siaya_Delta_Alive_130        wt,F2        wt,S1      F2,S1\n",
       "Siaya_Delta_Alive_131        F2,F2        wt,wt      F2,F2\n",
       "ML_17                        wt,wt        S3,S3      S3,S3\n",
       "ML_18                          ?,?          ?,?        ?,?\n",
       "mixedplate_negative            ?,?          ?,?        ?,?\n",
       "\n",
       "[672 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "kdr_origins = pd.concat([kdr_origin(gen_df.iloc[i]) for i in range(gen_df.shape[0])])\n",
    "kdr_origins['kdr_origin'] = kdr_origins.apply(\n",
    "    get_single_gen_call, axis = 1\n",
    ")\n",
    "\n",
    "kdr_origins"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d31864b4",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Merge kdr origins with metadata and write to file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "22f47c9a",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kdr_F_origin</th>\n",
       "      <th>kdr_S_origin</th>\n",
       "      <th>kdr_origin</th>\n",
       "      <th>taxon</th>\n",
       "      <th>location</th>\n",
       "      <th>country</th>\n",
       "      <th>plate</th>\n",
       "      <th>well_letter</th>\n",
       "      <th>well_number</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GH_01</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_02</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_03</th>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_04</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GH_05</th>\n",
       "      <td>F1,F1</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F1,F1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_130</th>\n",
       "      <td>wt,F2</td>\n",
       "      <td>wt,S1</td>\n",
       "      <td>F2,S1</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>Siaya</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Siaya_Delta_Alive_131</th>\n",
       "      <td>F2,F2</td>\n",
       "      <td>wt,wt</td>\n",
       "      <td>F2,F2</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>Siaya</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ML_17</th>\n",
       "      <td>wt,wt</td>\n",
       "      <td>S3,S3</td>\n",
       "      <td>S3,S3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ml</td>\n",
       "      <td>ml</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ML_18</th>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ml</td>\n",
       "      <td>ml</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mixedplate_negative</th>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "      <td>?,?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>random</td>\n",
       "      <td>random</td>\n",
       "      <td>7</td>\n",
       "      <td>H</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>672 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      kdr_F_origin kdr_S_origin kdr_origin    taxon location  \\\n",
       "GH_01                        F1,F1        wt,wt      F1,F1      NaN    Ghana   \n",
       "GH_02                        F1,F1        wt,wt      F1,F1      NaN    Ghana   \n",
       "GH_03                          ?,?          ?,?        ?,?      NaN    Ghana   \n",
       "GH_04                        F1,F1        wt,wt      F1,F1      NaN    Ghana   \n",
       "GH_05                        F1,F1        wt,wt      F1,F1      NaN    Ghana   \n",
       "...                            ...          ...        ...      ...      ...   \n",
       "Siaya_Delta_Alive_130        wt,F2        wt,S1      F2,S1  gambiae    Siaya   \n",
       "Siaya_Delta_Alive_131        F2,F2        wt,wt      F2,F2  gambiae    Siaya   \n",
       "ML_17                        wt,wt        S3,S3      S3,S3      NaN       ml   \n",
       "ML_18                          ?,?          ?,?        ?,?      NaN       ml   \n",
       "mixedplate_negative            ?,?          ?,?        ?,?      NaN   random   \n",
       "\n",
       "                      country  plate well_letter  well_number  latitude  \\\n",
       "GH_01                   Ghana      3           A            1       NaN   \n",
       "GH_02                   Ghana      3           A            2       NaN   \n",
       "GH_03                   Ghana      3           A            3       NaN   \n",
       "GH_04                   Ghana      3           A            4       NaN   \n",
       "GH_05                   Ghana      3           A            5       NaN   \n",
       "...                       ...    ...         ...          ...       ...   \n",
       "Siaya_Delta_Alive_130   Kenya      7           H            8       NaN   \n",
       "Siaya_Delta_Alive_131   Kenya      7           H            9       NaN   \n",
       "ML_17                      ml      7           H           10       NaN   \n",
       "ML_18                      ml      7           H           11       NaN   \n",
       "mixedplate_negative    random      7           H           12       NaN   \n",
       "\n",
       "                       longitude  \n",
       "GH_01                        NaN  \n",
       "GH_02                        NaN  \n",
       "GH_03                        NaN  \n",
       "GH_04                        NaN  \n",
       "GH_05                        NaN  \n",
       "...                          ...  \n",
       "Siaya_Delta_Alive_130        NaN  \n",
       "Siaya_Delta_Alive_131        NaN  \n",
       "ML_17                        NaN  \n",
       "ML_18                        NaN  \n",
       "mixedplate_negative          NaN  \n",
       "\n",
       "[672 rows x 11 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kdr_origins_df = pd.merge(kdr_origins, metadata, left_index = True, right_index = True)\n",
    "kdr_origins_df.to_csv('kdr_origins.csv', sep = '\\t')\n",
    "kdr_origins_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e0ce4db",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Create a table where each row is a haplotype instead of a genotype (although these are genotype-based calls, so order within each sample will be random, so they can be used for, say, mapping, but not haplotype clustering). Write this table to file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8ac2402e",
   "metadata": {
    "scrolled": true,
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kdr_origin</th>\n",
       "      <th>taxon</th>\n",
       "      <th>location</th>\n",
       "      <th>country</th>\n",
       "      <th>plate</th>\n",
       "      <th>well_letter</th>\n",
       "      <th>well_number</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Calvin_01</th>\n",
       "      <td>wt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Calvin_01</th>\n",
       "      <td>wt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Calvin_02</th>\n",
       "      <td>wt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Calvin_02</th>\n",
       "      <td>wt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Calvin_03</th>\n",
       "      <td>wt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>UNKN</td>\n",
       "      <td>14</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>random_funestus</th>\n",
       "      <td>F</td>\n",
       "      <td>funestus</td>\n",
       "      <td>random</td>\n",
       "      <td>random</td>\n",
       "      <td>7</td>\n",
       "      <td>E</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>random_stephensi</th>\n",
       "      <td>?</td>\n",
       "      <td>stephensi</td>\n",
       "      <td>random</td>\n",
       "      <td>random</td>\n",
       "      <td>7</td>\n",
       "      <td>F</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>random_stephensi</th>\n",
       "      <td>?</td>\n",
       "      <td>stephensi</td>\n",
       "      <td>random</td>\n",
       "      <td>random</td>\n",
       "      <td>7</td>\n",
       "      <td>F</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>random_stephensi2</th>\n",
       "      <td>?</td>\n",
       "      <td>stephensi2</td>\n",
       "      <td>random</td>\n",
       "      <td>random</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>random_stephensi2</th>\n",
       "      <td>?</td>\n",
       "      <td>stephensi2</td>\n",
       "      <td>random</td>\n",
       "      <td>random</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1344 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  kdr_origin       taxon location country  plate well_letter  \\\n",
       "Calvin_01                 wt         NaN     UNKN    UNKN     14           A   \n",
       "Calvin_01                 wt         NaN     UNKN    UNKN     14           A   \n",
       "Calvin_02                 wt         NaN     UNKN    UNKN     14           A   \n",
       "Calvin_02                 wt         NaN     UNKN    UNKN     14           A   \n",
       "Calvin_03                 wt         NaN     UNKN    UNKN     14           A   \n",
       "...                      ...         ...      ...     ...    ...         ...   \n",
       "random_funestus            F    funestus   random  random      7           E   \n",
       "random_stephensi           ?   stephensi   random  random      7           F   \n",
       "random_stephensi           ?   stephensi   random  random      7           F   \n",
       "random_stephensi2          ?  stephensi2   random  random      7           G   \n",
       "random_stephensi2          ?  stephensi2   random  random      7           G   \n",
       "\n",
       "                   well_number  latitude  longitude  \n",
       "Calvin_01                    1       NaN        NaN  \n",
       "Calvin_01                    1       NaN        NaN  \n",
       "Calvin_02                    2       NaN        NaN  \n",
       "Calvin_02                    2       NaN        NaN  \n",
       "Calvin_03                    3       NaN        NaN  \n",
       "...                        ...       ...        ...  \n",
       "random_funestus             12       NaN        NaN  \n",
       "random_stephensi            12       NaN        NaN  \n",
       "random_stephensi            12       NaN        NaN  \n",
       "random_stephensi2           12       NaN        NaN  \n",
       "random_stephensi2           12       NaN        NaN  \n",
       "\n",
       "[1344 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "kdr_genhap_origins_df = pd.DataFrame({'kdr_origin': ','.join(list(kdr_origins['kdr_origin'])).split(',')},\n",
    "                                     index = np.repeat(kdr_origins.index, 2)\n",
    ")\n",
    "kdr_genhap_origins_df = pd.merge(kdr_genhap_origins_df, metadata, left_index=True, right_index=True)\n",
    "kdr_genhap_origins_df.to_csv('kdr_genhap_origins.csv', sep = '\\t')\n",
    "kdr_genhap_origins_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dae4aae",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### Summarise results by taxon and location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6d3b20e5",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counts of origins:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kdr_origin</th>\n",
       "      <th>?</th>\n",
       "      <th>F</th>\n",
       "      <th>F1</th>\n",
       "      <th>F2</th>\n",
       "      <th>S</th>\n",
       "      <th>S1</th>\n",
       "      <th>S3</th>\n",
       "      <th>wt</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>taxon</th>\n",
       "      <th>location</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>arabiensis</th>\n",
       "      <th>random</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">coluzzii</th>\n",
       "      <th>random</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VK7</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>185</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>funestus</th>\n",
       "      <th>random</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gambiae</th>\n",
       "      <th>Siaya</th>\n",
       "      <td>94</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>283</td>\n",
       "      <td>21</td>\n",
       "      <td>103</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stephensi</th>\n",
       "      <th>random</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stephensi2</th>\n",
       "      <th>random</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "kdr_origin            ?  F   F1   F2   S   S1  S3  wt\n",
       "taxon      location                                  \n",
       "arabiensis random     0  0    0    0   0    0   0   2\n",
       "coluzzii   random     0  0    2    0   0    0   0   0\n",
       "           VK7        4  1  185    0   0    0   0   0\n",
       "funestus   random     0  2    0    0   0    0   0   0\n",
       "gambiae    Siaya     94  8    0  283  21  103   2  13\n",
       "stephensi  random     2  0    0    0   0    0   0   0\n",
       "stephensi2 random     2  0    0    0   0    0   0   0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Frequencies of known origins:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kdr_origin</th>\n",
       "      <th>F</th>\n",
       "      <th>F1</th>\n",
       "      <th>F2</th>\n",
       "      <th>S</th>\n",
       "      <th>S1</th>\n",
       "      <th>S3</th>\n",
       "      <th>wt</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>taxon</th>\n",
       "      <th>location</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>arabiensis</th>\n",
       "      <th>random</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">coluzzii</th>\n",
       "      <th>random</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VK7</th>\n",
       "      <td>0.0054</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>funestus</th>\n",
       "      <th>random</th>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gambiae</th>\n",
       "      <th>Siaya</th>\n",
       "      <td>0.0190</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.049</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.0047</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stephensi</th>\n",
       "      <th>random</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stephensi2</th>\n",
       "      <th>random</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "kdr_origin                F    F1    F2      S    S1      S3    wt\n",
       "taxon      location                                               \n",
       "arabiensis random    0.0000  0.00  0.00  0.000  0.00  0.0000  1.00\n",
       "coluzzii   random    0.0000  1.00  0.00  0.000  0.00  0.0000  0.00\n",
       "           VK7       0.0054  0.99  0.00  0.000  0.00  0.0000  0.00\n",
       "funestus   random    1.0000  0.00  0.00  0.000  0.00  0.0000  0.00\n",
       "gambiae    Siaya     0.0190  0.00  0.66  0.049  0.24  0.0047  0.03\n",
       "stephensi  random       NaN   NaN   NaN    NaN   NaN     NaN   NaN\n",
       "stephensi2 random       NaN   NaN   NaN    NaN   NaN     NaN   NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Count the number of occurances of each haplotypes in each population\n",
    "# \"values\" could be any column that isn't specified elsewhere in the function. But it's \n",
    "# not allowed to be blank, so we had to pick one. \n",
    "pop_origin_counts = pd.pivot_table(kdr_genhap_origins_df, \n",
    "                                   values = 'plate', \n",
    "                                   index = ['taxon', cohort_column], \n",
    "                                   columns = 'kdr_origin', \n",
    "                                   aggfunc = len\n",
    "                                  ).fillna(0).astype(int)\n",
    "\n",
    "# A function to round a number up to n_signif significant figures\n",
    "def signif(x, n_figs):\n",
    "    power = 10 ** np.floor(np.log10(np.abs(x).clip(1e-200)))\n",
    "    rounded = np.round(x / power, n_figs - 1) * power\n",
    "    return rounded\n",
    "\n",
    "# Calculate row totals of non-\"?\" columns\n",
    "row_totals = pop_origin_counts.drop('?', axis = 1).sum(axis = 1)\n",
    "\n",
    "# Calculate origin frequencies. We exclude the \"?\" calls for this\n",
    "pop_origin_freqs = pop_origin_counts.drop('?', axis = 1).div(row_totals, axis = 0)\n",
    "# Round to 2 significant figures\n",
    "pop_origin_freqs = signif(pop_origin_freqs, 2)\n",
    "print('Counts of origins:')\n",
    "display(pop_origin_counts)\n",
    "print('\\n\\nFrequencies of known origins:')\n",
    "display(pop_origin_freqs)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e782b7c6",
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "source": [
    "### That's all folks"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "malariagen_latest",
   "language": "python",
   "name": "malariagen_latest"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
